{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BIO INFORMATICS PART 2 : NETWORK MEDICINE PROJECT GROUP- 7\n",
    "-  VIKRANTH ALE [1873995]\n",
    "-  KENTARO KATO [1851049]\n",
    "\n",
    "\n",
    "#### Disgenet Pathology Code: C0026998\n",
    "#### Disgenet Pathology Name: Acute Myeloid Leukemia, M1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"Loading required libraries \"\"\"\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from tqdm import trange,tqdm_notebook\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 a) Explore the DisGeNet dataset, find the disease of interest and get the list of human genes involved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of observations in our dataset (107, 18)\n",
      "\n",
      "The Disease we are interested for our analysis is Acute Myeloid Leukemia, M1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Disease_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Gene_Full_Name</th>\n",
       "      <th>Protein_Class</th>\n",
       "      <th>N_diseases_g</th>\n",
       "      <th>DSI_g</th>\n",
       "      <th>DPI_g</th>\n",
       "      <th>pLI</th>\n",
       "      <th>Score_gda</th>\n",
       "      <th>EL_gda</th>\n",
       "      <th>EI_gda</th>\n",
       "      <th>N_PMIDs</th>\n",
       "      <th>N_SNPs_gda</th>\n",
       "      <th>First_Ref</th>\n",
       "      <th>Last_Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acute Myeloid Leukemia, M1</td>\n",
       "      <td>C0026998</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>2322</td>\n",
       "      <td>P36888</td>\n",
       "      <td>fms related tyrosine kinase 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21</td>\n",
       "      <td>0.503</td>\n",
       "      <td>0.517</td>\n",
       "      <td>6.136100e-01</td>\n",
       "      <td>0.52</td>\n",
       "      <td>no reported evidence</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2002.0</td>\n",
       "      <td>2017.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acute Myeloid Leukemia, M1</td>\n",
       "      <td>C0026998</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>4869</td>\n",
       "      <td>P06748</td>\n",
       "      <td>nucleophosmin 1</td>\n",
       "      <td>chaperone</td>\n",
       "      <td>13</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.690</td>\n",
       "      <td>9.983100e-01</td>\n",
       "      <td>0.50</td>\n",
       "      <td>no reported evidence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>2017.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acute Myeloid Leukemia, M1</td>\n",
       "      <td>C0026998</td>\n",
       "      <td>IDH1</td>\n",
       "      <td>3417</td>\n",
       "      <td>O75874</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 1, cytosolic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.690</td>\n",
       "      <td>3.337500e-12</td>\n",
       "      <td>0.32</td>\n",
       "      <td>no reported evidence</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2017.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acute Myeloid Leukemia, M1</td>\n",
       "      <td>C0026998</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>4297</td>\n",
       "      <td>Q03164</td>\n",
       "      <td>lysine methyltransferase 2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>44</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.793</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>no reported evidence</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>2016.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acute Myeloid Leukemia, M1</td>\n",
       "      <td>C0026998</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>3418</td>\n",
       "      <td>P48735</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 2, mitochon...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.655</td>\n",
       "      <td>8.652500e-01</td>\n",
       "      <td>0.31</td>\n",
       "      <td>no reported evidence</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2017.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Disease Disease_id   Gene  Gene_id UniProt  \\\n",
       "0  Acute Myeloid Leukemia, M1   C0026998   FLT3     2322  P36888   \n",
       "1  Acute Myeloid Leukemia, M1   C0026998   NPM1     4869  P06748   \n",
       "2  Acute Myeloid Leukemia, M1   C0026998   IDH1     3417  O75874   \n",
       "3  Acute Myeloid Leukemia, M1   C0026998  KMT2A     4297  Q03164   \n",
       "4  Acute Myeloid Leukemia, M1   C0026998   IDH2     3418  P48735   \n",
       "\n",
       "                                      Gene_Full_Name Protein_Class  \\\n",
       "0                      fms related tyrosine kinase 3           NaN   \n",
       "1                                    nucleophosmin 1     chaperone   \n",
       "2    isocitrate dehydrogenase (NADP(+)) 1, cytosolic           NaN   \n",
       "3                        lysine methyltransferase 2A           NaN   \n",
       "4  isocitrate dehydrogenase (NADP(+)) 2, mitochon...           NaN   \n",
       "\n",
       "   N_diseases_g  DSI_g  DPI_g           pLI  Score_gda                EL_gda  \\\n",
       "0            21  0.503  0.517  6.136100e-01       0.52  no reported evidence   \n",
       "1            13  0.496  0.690  9.983100e-01       0.50  no reported evidence   \n",
       "2            48  0.463  0.690  3.337500e-12       0.32  no reported evidence   \n",
       "3            44  0.460  0.793  1.000000e+00       0.32  no reported evidence   \n",
       "4            92  0.469  0.655  8.652500e-01       0.31  no reported evidence   \n",
       "\n",
       "   EI_gda  N_PMIDs  N_SNPs_gda  First_Ref  Last_Ref  \n",
       "0     1.0       13           0     2002.0    2017.0  \n",
       "1     NaN        4           0     2010.0    2017.0  \n",
       "2     1.0        2           0     2011.0    2017.0  \n",
       "3     1.0        2           0     1996.0    2016.0  \n",
       "4     1.0        2           0     2011.0    2017.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Curated gene-disease associations data from DisGeNet with corresponding Uniprot entries\n",
    "\n",
    "df = pd.read_csv(\"C0026998_disease_gda_summary_CURATED.tsv\", sep='\\t')\n",
    "df.to_csv('Acute_Myeloid_Leukemia_M1.csv', encoding='utf8',index=False)\n",
    "\n",
    "print('Total number of observations in our dataset {}'.format(df.shape))\n",
    "\n",
    "print('')\n",
    "#### Disease of Interest\n",
    "\n",
    "print(\"The Disease we are interested for our analysis is {}\".format(df['Disease'][0]))\n",
    "\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes involved: 107\n",
      "Number of UniProtAC involved: 106\n",
      "Number of null values in UniProt is 1\n",
      "GeneSymbol for which uniprot entry is null: ['DLEU2']\n"
     ]
    }
   ],
   "source": [
    "# basic stats about our dataset\n",
    "\n",
    "print('Number of genes involved: {}'.format(df['Gene'].nunique()))\n",
    "print('Number of UniProtAC involved: {}'.format(df['UniProt'].nunique()))\n",
    "\n",
    "print('Number of null values in UniProt is {}'.format(df['UniProt'].isnull().sum())) # \n",
    "print('GeneSymbol for which uniprot entry is null: {}'.format(df[df['UniProt'].isnull()]['Gene'].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# removing the gene that has not no uniprot entry as it is typically found in Mouse\n",
    "df = df[df.Gene != 'DLEU2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>UniProtAC</th>\n",
       "      <th>ProteinName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>2322</td>\n",
       "      <td>P36888</td>\n",
       "      <td>fms related tyrosine kinase 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>4869</td>\n",
       "      <td>P06748</td>\n",
       "      <td>nucleophosmin 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IDH1</td>\n",
       "      <td>3417</td>\n",
       "      <td>O75874</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 1, cytosolic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KMT2A</td>\n",
       "      <td>4297</td>\n",
       "      <td>Q03164</td>\n",
       "      <td>lysine methyltransferase 2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>3418</td>\n",
       "      <td>P48735</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 2, mitochon...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GeneSymbol  GeneID UniProtAC  \\\n",
       "0       FLT3    2322    P36888   \n",
       "1       NPM1    4869    P06748   \n",
       "2       IDH1    3417    O75874   \n",
       "3      KMT2A    4297    Q03164   \n",
       "4       IDH2    3418    P48735   \n",
       "\n",
       "                                         ProteinName  \n",
       "0                      fms related tyrosine kinase 3  \n",
       "1                                    nucleophosmin 1  \n",
       "2    isocitrate dehydrogenase (NADP(+)) 1, cytosolic  \n",
       "3                        lysine methyltransferase 2A  \n",
       "4  isocitrate dehydrogenase (NADP(+)) 2, mitochon...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dff = df[['Gene','Gene_id','UniProt','Gene_Full_Name']]\n",
    "dff.rename(columns={'Gene':'GeneSymbol','UniProt':'UniProtAC','Gene_id':'GeneID','Gene_Full_Name':'ProteinName'},inplace=True)\n",
    "dff.to_csv('gene_map.csv',encoding='utf8',index=False)\n",
    "dff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### checking the updated symbols and approved seed genes data as per HGNC website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106\n"
     ]
    }
   ],
   "source": [
    "gene_list = dff['GeneSymbol']\n",
    "print(len(gene_list))\n",
    "gene_list.to_csv('gene_list.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(112, 6)\n",
      "The number of approved symbol: 103\n",
      "The number of symbols not approved: 3\n",
      "\n",
      " Here's the list of genes that symbols are not approved\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>Match type</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>SEPT9</td>\n",
       "      <td>Previous symbol</td>\n",
       "      <td>SEPTIN9</td>\n",
       "      <td>septin 9</td>\n",
       "      <td>HGNC:7323</td>\n",
       "      <td>17q25.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>Previous symbol</td>\n",
       "      <td>H1-0</td>\n",
       "      <td>H1.0 linker histone</td>\n",
       "      <td>HGNC:4714</td>\n",
       "      <td>22q13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>Previous symbol</td>\n",
       "      <td>H1-2</td>\n",
       "      <td>H1.2 linker histone, cluster member</td>\n",
       "      <td>HGNC:4716</td>\n",
       "      <td>6p22.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Input       Match type Approved symbol  \\\n",
       "53     SEPT9  Previous symbol         SEPTIN9   \n",
       "86      H1F0  Previous symbol            H1-0   \n",
       "87  HIST1H1C  Previous symbol            H1-2   \n",
       "\n",
       "                          Approved name    HGNC ID Location  \n",
       "53                             septin 9  HGNC:7323  17q25.3  \n",
       "86                  H1.0 linker histone  HGNC:4714  22q13.1  \n",
       "87  H1.2 linker histone, cluster member  HGNC:4716   6p22.2  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc = pd.read_csv('hgnc-symbol-check1.csv',skiprows=1)\n",
    "print(hgnc.shape)\n",
    "print('The number of approved symbol: {}'.format(len(hgnc[hgnc['Match type'] == 'Approved symbol'])))\n",
    "print('The number of symbols not approved: {}'.format(len(hgnc[hgnc['Match type'] == 'Previous symbol'])))\n",
    "\n",
    "print(\"\\n Here's the list of genes that symbols are not approved\")\n",
    "hgnc[hgnc['Match type'] == 'Previous symbol'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(106, 4)\n"
     ]
    }
   ],
   "source": [
    "non_approved_genes = hgnc[hgnc['Match type']=='Previous symbol']['Input'].values\n",
    "approved_name = hgnc[hgnc['Match type']=='Previous symbol']['Approved symbol'].values\n",
    "\n",
    "for i in range(len(non_approved_genes)):\n",
    "    dff = dff.replace(non_approved_genes[i],approved_name[i])\n",
    "\n",
    "print(dff.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106\n"
     ]
    }
   ],
   "source": [
    "our_uniprot = dff['UniProtAC']\n",
    "print(len(our_uniprot))\n",
    "our_uniprot.to_csv('our_uniprot', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 b) Collecting required seed genes information from UniProt website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(106, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtAC</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>ProteinName</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P36888</td>\n",
       "      <td>FLT3_HUMAN</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>Receptor-type tyrosine-protein kinase FLT3</td>\n",
       "      <td>Tyrosine-protein kinase that acts as cell-sur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P06748</td>\n",
       "      <td>NPM_HUMAN</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>Nucleophosmin</td>\n",
       "      <td>Involved in diverse cellular processes such a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O75874</td>\n",
       "      <td>IDHC_HUMAN</td>\n",
       "      <td>IDH1</td>\n",
       "      <td>Isocitrate dehydrogenase [NADP] cytoplasmic</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q03164</td>\n",
       "      <td>KMT2A_HUMAN</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>Histone-lysine N-methyltransferase 2A</td>\n",
       "      <td>Histone methyltransferase that plays an essen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P48735</td>\n",
       "      <td>IDHP_HUMAN</td>\n",
       "      <td>IDH2</td>\n",
       "      <td>Isocitrate dehydrogenase [NADP], mitochondrial</td>\n",
       "      <td>Plays a role in intermediary metabolism and e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniProtAC   Entry name GeneSymbol  \\\n",
       "0    P36888   FLT3_HUMAN       FLT3   \n",
       "1    P06748    NPM_HUMAN       NPM1   \n",
       "2    O75874   IDHC_HUMAN       IDH1   \n",
       "3    Q03164  KMT2A_HUMAN      KMT2A   \n",
       "4    P48735   IDHP_HUMAN       IDH2   \n",
       "\n",
       "                                       ProteinName  \\\n",
       "0      Receptor-type tyrosine-protein kinase FLT3    \n",
       "1                                   Nucleophosmin    \n",
       "2     Isocitrate dehydrogenase [NADP] cytoplasmic    \n",
       "3           Histone-lysine N-methyltransferase 2A    \n",
       "4  Isocitrate dehydrogenase [NADP], mitochondrial    \n",
       "\n",
       "                                         Description  \n",
       "0   Tyrosine-protein kinase that acts as cell-sur...  \n",
       "1   Involved in diverse cellular processes such a...  \n",
       "2                                                nan  \n",
       "3   Histone methyltransferase that plays an essen...  \n",
       "4   Plays a role in intermediary metabolism and e...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#raw file from uniprot website using our gene_list\n",
    "uniprot = pd.read_csv('uniprot-yourlist_M20200108216DA2B77BFBD2E6699CA9B6D1C41EB26F69CBG.tab',sep='\\t')\n",
    "#pre-processing\n",
    "uniprot.replace({'Function [CC]':'FUNCTION:'},'',regex=True,inplace=True)\n",
    "uniprot.rename(columns={'Entry':'UniProtAC','Function [CC]':'Description','Gene names  (primary )':'GeneSymbol','Protein names':'ProteinName'},inplace=True)\n",
    "uniprot['ProteinName'] = uniprot['ProteinName'].str.split('(').str[0]\n",
    "uniprot['Description'] = uniprot['Description'].astype(str)\n",
    "uniprot['Description'] = uniprot['Description'].str.split('.').str[0]\n",
    "uniprot.to_csv('uniprot_final.csv',encoding='utf8',index=False)\n",
    "print(uniprot.shape)\n",
    "uniprot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 c) Store the data gathered in a table in an easily accessible format of your choice (csv, tab, excel, etc)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>UniProtAC</th>\n",
       "      <th>ProteinName</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>2322</td>\n",
       "      <td>P36888</td>\n",
       "      <td>fms related tyrosine kinase 3</td>\n",
       "      <td>Tyrosine-protein kinase that acts as cell-sur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>4869</td>\n",
       "      <td>P06748</td>\n",
       "      <td>nucleophosmin 1</td>\n",
       "      <td>Involved in diverse cellular processes such a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IDH1</td>\n",
       "      <td>3417</td>\n",
       "      <td>O75874</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 1, cytosolic</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KMT2A</td>\n",
       "      <td>4297</td>\n",
       "      <td>Q03164</td>\n",
       "      <td>lysine methyltransferase 2A</td>\n",
       "      <td>Histone methyltransferase that plays an essen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>3418</td>\n",
       "      <td>P48735</td>\n",
       "      <td>isocitrate dehydrogenase (NADP(+)) 2, mitochon...</td>\n",
       "      <td>Plays a role in intermediary metabolism and e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GeneSymbol  GeneID UniProtAC  \\\n",
       "0       FLT3    2322    P36888   \n",
       "1       NPM1    4869    P06748   \n",
       "2       IDH1    3417    O75874   \n",
       "3      KMT2A    4297    Q03164   \n",
       "4       IDH2    3418    P48735   \n",
       "\n",
       "                                         ProteinName  \\\n",
       "0                      fms related tyrosine kinase 3   \n",
       "1                                    nucleophosmin 1   \n",
       "2    isocitrate dehydrogenase (NADP(+)) 1, cytosolic   \n",
       "3                        lysine methyltransferase 2A   \n",
       "4  isocitrate dehydrogenase (NADP(+)) 2, mitochon...   \n",
       "\n",
       "                                         Description  \n",
       "0   Tyrosine-protein kinase that acts as cell-sur...  \n",
       "1   Involved in diverse cellular processes such a...  \n",
       "2                                                nan  \n",
       "3   Histone methyltransferase that plays an essen...  \n",
       "4   Plays a role in intermediary metabolism and e...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final = pd.merge(dff[['GeneSymbol', 'GeneID', 'UniProtAC', 'ProteinName']],\n",
    "                    uniprot[['UniProtAC', 'GeneSymbol', 'Description']],\n",
    "                   on=['GeneSymbol','UniProtAC'])\n",
    "\n",
    "df_final.to_csv('df_final.csv', encoding='utf8', index=False)\n",
    "print(len(df_final))\n",
    "df_final.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of Seed Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106\n",
      "['ADCY7', 'AGRN', 'ANXA2', 'ANXA4', 'ANXA5', 'ANXA6', 'AQP9', 'ASMTL', 'ASXL2', 'ATP1B1', 'BAALC', 'BACH2', 'BCL2', 'CAPG', 'CAPN2', 'CASP7', 'CBFB', 'CCND2', 'CD33', 'CD44', 'CD9', 'CDK6', 'CEBPA', 'CEBPD', 'CHMP5', 'CNR2', 'CSF1R', 'CSF2', 'CSF3', 'CST3', 'CTNNA1', 'CTSH', 'CTSZ', 'DAPK1', 'DHX15', 'DNMT3A', 'EHD3', 'EHMT2', 'EIF4EBP1', 'ENAH', 'ENO2', 'ERG', 'FAS', 'FHL2', 'FLT3', 'FOXO1', 'FXYD6', 'GAS2L1', 'GATA2', 'GFI1', 'GTF2I', 'H1-0', 'H1-2', 'HGF', 'HOXA9', 'HSPB1', 'ID2', 'IDH1', 'IDH2', 'IFI30', 'JAK2', 'KMT2A', 'KMT2C', 'KRAS', 'LPAR1', 'LYL1', 'MET', 'MN1', 'MX1', 'MYC', 'MYH11', 'NF1', 'NPM1', 'NRAS', 'NUP98', 'PDE4B', 'PIM2', 'POU4F1', 'PSIP1', 'PTPN11', 'PXDN', 'RASGRP1', 'RGS2', 'RUNX1', 'RUNX1T1', 'RUNX3', 'S100A10', 'S100A8', 'SEPTIN9', 'SGK1', 'SPARC', 'SPI1', 'SPRY4', 'STAT3', 'SVIL', 'SYNGR1', 'TCEA2', 'TNFSF10', 'TRH', 'TRIO', 'TSC2', 'TUBB2A', 'VOPP1', 'VSIG4', 'WT1', 'ZBTB7A']\n"
     ]
    }
   ],
   "source": [
    "seed_genes = list(df_final['GeneSymbol'].unique())\n",
    "print(len(seed_genes))\n",
    "print(sorted(seed_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of Seed Gene IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106\n",
      "[2322, 4869, 3417, 4297, 3418, 4629, 2624, 1050, 6281, 6446, 6678, 6688, 4763, 6774, 6840, 6919, 7200, 6279, 5997, 4330, 4599, 4609, 55252, 4893, 4928, 5142, 5457, 5781, 7204, 7249, 11326, 30845, 51341, 51510, 53826, 55740, 58508, 60468, 79870, 81552, 81848, 11168, 11040, 10919, 7280, 7490, 7837, 8623, 8743, 9145, 10125, 10437, 10634, 10801, 375790, 113, 864, 865, 894, 928, 945, 960, 1021, 1052, 1269, 1436, 862, 861, 840, 302, 307, 308, 309, 355, 366, 481, 596, 822, 824, 1437, 1440, 1471, 2969, 3005, 3006, 3082, 3205, 3315, 3398, 3717, 3845, 4066, 2672, 2308, 2274, 1495, 1512, 1522, 1612, 1665, 1788, 1902, 1978, 2026, 2078, 4233]\n"
     ]
    }
   ],
   "source": [
    "seed_geneIDs = list(df_final['GeneID'].unique())\n",
    "print(len(seed_genes))\n",
    "print((seed_geneIDs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 COLLECT INTERACTION DATA\n",
    "- a) Downloading the data from BioGrid database\n",
    "- b) Store the data gathered from the two DBs in two different tables/matrices in an easily accessible format of your choice.\n",
    "- c) Summarize the main results in a table reporting\n",
    "\n",
    "- [Download data from BioGrid Database](https://downloads.thebiogrid.org//BioGRID//Latest-Release// \"BioGrid Database\")\n",
    "\n",
    "- FILE NAME : BIOGRID-ALL-LATEST.tab2.zip\n",
    "\n",
    "<font color='green'> **Loading and cleaning BioGrid database</font>**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entire dataset\n",
      "(1739944, 24)\n",
      "Only Human genes dataset\n",
      "(493288, 24)\n",
      "Select only required features\n",
      "(493288, 2)\n",
      "Convert gene symbols to approved ones ------------------------------------------------\n",
      "The number of all genes found in biogrid:  17913\n",
      "The number of approved symbol: 17644\n",
      "The number of not approved symbol: 8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:00<00:00, 11.53it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\" Loading the Biogrid database from the file and extracting the required columns \"\"\"\n",
    "df_biogrid = pd.read_csv('BIOGRID-ALL-3.5.180.tab2.txt', sep='\\t')\n",
    "print('Entire dataset')\n",
    "print(df_biogrid.shape)\n",
    "\n",
    "# extract only genes human-human interactions\n",
    "df_biogrid = df_biogrid.loc[(df_biogrid['Organism Interactor A'] == 9606) & (df_biogrid['Organism Interactor B'] == 9606)]\n",
    "print('Only Human genes dataset')\n",
    "print(df_biogrid.shape)\n",
    "\n",
    "# reduce and rename the columns while resetting the index\n",
    "df_biogrid = df_biogrid[['Official Symbol Interactor A',\n",
    "                         'Official Symbol Interactor B']]\n",
    "\n",
    "df_biogrid.rename(columns={'Official Symbol Interactor A':'GeneA',\n",
    "                           'Official Symbol Interactor B':'GeneB'}, inplace=True)\n",
    "\n",
    "df_biogrid.reset_index(inplace=True)\n",
    "df_biogrid = df_biogrid.drop(columns=['index'])\n",
    "\n",
    "\n",
    "print('Select only required features')\n",
    "print(df_biogrid.shape)\n",
    "\n",
    "\n",
    "print('Convert gene symbols to approved ones ------------------------------------------------')\n",
    "\n",
    "all_genes_biogrid = list(set(df_biogrid['GeneA'].values)) + list(set(df_biogrid['GeneB'].values))\n",
    "all_genes_biogrid = list(set(all_genes_biogrid))\n",
    "print('The number of all genes found in biogrid: ', len(all_genes_biogrid))\n",
    "all_genes_biogrid = pd.Series(all_genes_biogrid)\n",
    "all_genes_biogrid.to_csv('1.2. genes_biogrid', index = False)\n",
    "\n",
    "hgnc = pd.read_csv('1.2. hgnc-symbol-check_biogrid.csv', skiprows=1)\n",
    "print('The number of approved symbol: {}'.format(len(hgnc[hgnc['Match type'] == 'Approved symbol'])))\n",
    "print('The number of not approved symbol: {}'.format(len(hgnc[hgnc['Match type'] == 'Previous symbol'])))\n",
    "\n",
    "non_approved_genes = hgnc[hgnc['Match type'] == 'Previous symbol']['Input'].values\n",
    "approved_name = hgnc[hgnc['Match type'] == 'Previous symbol']['Approved symbol'].values\n",
    "\n",
    "\n",
    "# change gene symbols to approved ones\n",
    "for i in trange(len(non_approved_genes)):\n",
    "    df_biogrid = df_biogrid.replace(non_approved_genes[i], approved_name[i])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of seed genes found in Biogrid "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No. of all genes found in this database: 17909\n",
      "No. of seed genes found: 100\n",
      "Genes not found in the database: {'H1-2', 'H1-0', 'TRH', 'CSF3', 'CNR2', 'SEPTIN9'}\n"
     ]
    }
   ],
   "source": [
    "# Gathering genes data in the BioGrid database\n",
    "\n",
    "As = list(df_biogrid.GeneA.values)\n",
    "Bs = list(df_biogrid.GeneB.values)\n",
    "all_genes_biogrid = set(As + Bs)\n",
    "print('No. of all genes found in this database: {}'.format(len(all_genes_biogrid)))\n",
    "\n",
    "seed_genes_biogrid = list(all_genes_biogrid.intersection(set(seed_genes)))\n",
    "\n",
    "print('No. of seed genes found: {}'.format(len(seed_genes_biogrid)))\n",
    "print('Genes not found in the database: {}'.format(set(seed_genes).difference(all_genes_biogrid)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total Number of interacting proteins including seed genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'>**Interactions among the seed genes **</font>\n",
    "\n",
    "- GeneA AND GeneB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Number of interactions between the seed genes 318\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>KMT2A</td>\n",
       "      <td>KMT2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1287</th>\n",
       "      <td>ASMTL</td>\n",
       "      <td>ASMTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3888</th>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>MYC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3889</th>\n",
       "      <td>MYC</td>\n",
       "      <td>DNMT3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3890</th>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>MYC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneA   GeneB\n",
       "193    KMT2A   KMT2A\n",
       "1287   ASMTL   ASMTL\n",
       "3888  DNMT3A     MYC\n",
       "3889     MYC  DNMT3A\n",
       "3890  DNMT3A     MYC"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of interactions amongst the seed genes (Among blue nodes and blue line in the diagram)\n",
    "\n",
    "df_biogrid_seed_genes = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) \n",
    "                                       & df_biogrid['GeneB'].isin(seed_genes_biogrid)]\n",
    "print(\"The Number of interactions between the seed genes\",len(df_biogrid_seed_genes))\n",
    "df_biogrid_seed_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'>**Interactions with atleast one seed gene or a non seed Gene**</font>\n",
    "\n",
    "- GeneA OR GeneB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Number of interactions between the seed genes or Non seed genes 16570\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CSF1R</td>\n",
       "      <td>GRB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>KDR</td>\n",
       "      <td>ANXA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SUMO1</td>\n",
       "      <td>FAS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneA  GeneB\n",
       "3   GATA2    PML\n",
       "4    RPA2  STAT3\n",
       "15  CSF1R   GRB2\n",
       "80    KDR  ANXA5\n",
       "87  SUMO1    FAS"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of interactions with atleast one seed genes or non seed genes\n",
    "#(Between blue and yellow nodes with blue line in the diagram)\n",
    "\n",
    "df_biogrid_seed_non_seed = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) \n",
    "                                          | df_biogrid['GeneB'].isin(seed_genes_biogrid)]\n",
    "print(\"The Number of interactions between the seed genes or Non seed genes\",len(df_biogrid_seed_non_seed))\n",
    "df_biogrid_seed_non_seed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'> ** Getting the Number of Non seed Genes**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of seed genes: 100\n",
      "\n",
      "Number of non seed genes present in biogrid are  5826\n",
      "\n",
      "Total no. of interacting genes (seed genes and non seed genes): 5926\n"
     ]
    }
   ],
   "source": [
    "print('The number of seed genes: {}'.format(len(seed_genes_biogrid)))\n",
    "print('')\n",
    "\n",
    "df_biogrid_seed = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) | df_biogrid['GeneB'].isin(seed_genes_biogrid)]\n",
    "\n",
    "#Number of non seed genes (red dotted line in the diagram)\n",
    "non_seed_genesA = set(df_biogrid_seed['GeneA']).difference(set(seed_genes))\n",
    "non_seed_genesB = set(df_biogrid_seed['GeneB']).difference(set(seed_genes))\n",
    "\n",
    "non_seed_genes_biogrid  = list(non_seed_genesA.union(non_seed_genesB))\n",
    "print(\"Number of non seed genes present in biogrid are \",len(non_seed_genes_biogrid))\n",
    "print('')\n",
    "all_genes = non_seed_genes_biogrid + seed_genes_biogrid\n",
    "print('Total no. of interacting genes (seed genes and non seed genes): {}'.format(len(all_genes)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'>**Interactions among Non seed Genes **</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #Number of interactions between the non seed genes (red dotted line in the diagram)\n",
    "# df_biogrid_non_seed = df_biogrid.loc[df_biogrid['GeneA'].isin(non_seed_genes_biogrid) \n",
    "#                                      & df_biogrid['GeneB'].isin(non_seed_genes_biogrid)]\n",
    "# print(\"The Number of interactions between the non seed genes A & B :\",len(df_biogrid_non_seed))\n",
    "# df_biogrid_non_seed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Number of interactions  A & B : 269952\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ARF1</td>\n",
       "      <td>GGA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>XRN1</td>\n",
       "      <td>ALDOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>APP</td>\n",
       "      <td>APPBP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneA   GeneB\n",
       "3  GATA2     PML\n",
       "4   RPA2   STAT3\n",
       "5   ARF1    GGA3\n",
       "8   XRN1   ALDOA\n",
       "9    APP  APPBP2"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Number of interactions between the non seed genes (red dotted line in the diagram)\n",
    "df_biogrid_non_seed = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) | df_biogrid['GeneB'].isin(seed_genes_biogrid) |\\\n",
    "                                    (df_biogrid['GeneA'].isin(non_seed_genes_biogrid) & df_biogrid['GeneB'].isin(non_seed_genes_biogrid))]\n",
    "                                     \n",
    "print(\"Total Number of interactions  A & B :\",len(df_biogrid_non_seed))\n",
    "df_biogrid_non_seed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Number of interactions  A & B : 269952\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ARF1</td>\n",
       "      <td>GGA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>XRN1</td>\n",
       "      <td>ALDOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>APP</td>\n",
       "      <td>APPBP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneA   GeneB\n",
       "3  GATA2     PML\n",
       "4   RPA2   STAT3\n",
       "5   ARF1    GGA3\n",
       "8   XRN1   ALDOA\n",
       "9    APP  APPBP2"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Number of interactions between the non seed genes (red dotted line in the diagram)\n",
    "df_biogrid_final = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) | df_biogrid['GeneB'].isin(seed_genes_biogrid) |\\\n",
    "                                    (df_biogrid['GeneA'].isin(non_seed_genes_biogrid) & df_biogrid['GeneB'].isin(non_seed_genes_biogrid))]\n",
    "                                     \n",
    "print(\"Total Number of interactions  A & B :\",len(df_biogrid_final))\n",
    "df_biogrid_final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneA  GeneB\n",
       "0  MAP2K4   FLNC\n",
       "1    MYPN  ACTN2\n",
       "2   ACVR1   FNTA\n",
       "3   GATA2    PML\n",
       "4    RPA2  STAT3"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Storing all the BioGrid interactions data into an easily accessible excel sheet format.\n",
    "\n",
    "writer_object = pd.ExcelWriter('master_df_biogrid.xlsx',engine='xlsxwriter')\n",
    "\n",
    "df_biogrid.to_excel(writer_object,sheet_name='df_biogrid',encoding='utf8',index=False)\n",
    "df_biogrid_seed_genes.to_excel(writer_object,sheet_name='df_biogrid_seed_genes',encoding='utf8',index=False)\n",
    "df_biogrid_seed_non_seed.to_excel(writer_object,sheet_name='df_biogrid_seed_non_seed',encoding='utf8',index=False)\n",
    "df_biogrid_non_seed.to_excel(writer_object,sheet_name='df_biogrid_non_seed',encoding='utf8',index=False)\n",
    "#df_biogrid_atleast_one_non_seed.to_excel(writer_object,sheet_name='df_biogrid_atleast_one_non_seed',encoding='utf8',index=False)\n",
    "\n",
    "writer_object.save()\n",
    "\n",
    "df_biogrid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Repeating the above steps for Seed and Non-seed genes in IID Database \n",
    "\n",
    "- Getting IID data for the seed genes from the below link\n",
    "- [Download from IID Database](http://iid.ophid.utoronto.ca/search_by_proteins/)\n",
    "- Interactions in IID dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### IID Pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8764, 4)\n",
      "Convert gene symbols to approved ones ------------------------------------------------\n",
      "The number of all genes found in IID:  4773\n",
      "The number of approved symbol: 4669\n",
      "The number of not approved symbol: 126\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P49411</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>TUFM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q14697</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>GANAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q99759</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAP3K3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniProtA UniProtB GeneA   GeneB\n",
       "0   O00141   P49411  SGK1    TUFM\n",
       "1   O00141   Q14697  SGK1   GANAB\n",
       "2   O00141   Q99759  SGK1  MAP3K3"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_iid = pd.read_csv('PPIs.txt',sep='\\t')\n",
    "df_iid = df_iid[['UniProt1','UniProt2','symbol1','symbol2']]\n",
    "df_iid['symbol1'] = df_iid['symbol1'].str.split(';').str[0]\n",
    "df_iid['symbol1'] = df_iid['symbol1'].str.split(' ').str[0]\n",
    "df_iid['symbol2'] = df_iid['symbol2'].str.split(';').str[0]\n",
    "df_iid['symbol2'] = df_iid['symbol2'].str.split(' ').str[0]\n",
    "df_iid.to_csv('IIDdata.csv',encoding='utf8',index=False)\n",
    "\n",
    "df_iid.rename(columns={ 'symbol1':'GeneA',\n",
    "                                          'symbol2':'GeneB',\n",
    "                                          'UniProt1':'UniProtA',\n",
    "                                          'UniProt2':'UniProtB'}, inplace=True)\n",
    "df_iid.reset_index(inplace=True)\n",
    "df_iid = df_iid.drop(columns=['index'])\n",
    "\n",
    "print(df_iid.shape)\n",
    "\n",
    "print('Convert gene symbols to approved ones ------------------------------------------------')\n",
    "\n",
    "all_genes_iid = list(set(df_iid['GeneA'].values)) + list(set(df_iid['GeneB'].values))\n",
    "all_genes_iid = list(set(all_genes_iid))\n",
    "print('The number of all genes found in IID: ', len(all_genes_iid))\n",
    "all_genes_iid = pd.Series(all_genes_iid)\n",
    "all_genes_iid.to_csv('1.2. genes_iid', index = False)\n",
    "\n",
    "hgnc = pd.read_csv('1.2. hgnc-symbol-check_iid.csv', skiprows=1)\n",
    "print('The number of approved symbol: {}'.format(len(hgnc[hgnc['Match type'] == 'Approved symbol'])))\n",
    "print('The number of not approved symbol: {}'.format(len(hgnc[hgnc['Match type'] == 'Previous symbol'])))\n",
    "\n",
    "non_approved_genes = hgnc[hgnc['Match type'] == 'Previous symbol']['Input'].values\n",
    "approved_name = hgnc[hgnc['Match type'] == 'Previous symbol']['Approved symbol'].values\n",
    "\n",
    "for i in range(len(non_approved_genes)):\n",
    "    df_iid = df_iid.replace(non_approved_genes[i], approved_name[i])\n",
    "    \n",
    "df_iid.to_csv('our_iid.csv', encoding='utf8', index=False)\n",
    "\n",
    "df_iid.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total No. of genes found in IID database: 4759\n",
      "No. of seed genes found in IID : 106\n",
      "Genes not found in IID database: set()\n"
     ]
    }
   ],
   "source": [
    "# Getting the count of all seed genes present in IID database\n",
    "\n",
    "iid_A = list(df_iid.GeneA.values)\n",
    "iid_B = list(df_iid.GeneB.values)\n",
    "\n",
    "all_iid_genes = set(iid_A+iid_B)\n",
    "\n",
    "print('Total No. of genes found in IID database: {}'.format(len(all_iid_genes)))\n",
    "\n",
    "seed_genes_iid = list(all_iid_genes.intersection(set(seed_genes)))\n",
    "\n",
    "print('No. of seed genes found in IID : {}'.format(len(seed_genes_iid)))\n",
    "print('Genes not found in IID database: {}'.format(set(seed_genes).difference(all_iid_genes)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'>**Seed Genes Interactions in IID**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Number of interactions between the seed genes 166\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P49815</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>TSC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>O00141</td>\n",
       "      <td>O00141</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>SGK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>O43143</td>\n",
       "      <td>P01106</td>\n",
       "      <td>DHX15</td>\n",
       "      <td>MYC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>O43143</td>\n",
       "      <td>P06748</td>\n",
       "      <td>DHX15</td>\n",
       "      <td>NPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>O60674</td>\n",
       "      <td>P78347</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>GTF2I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProtA UniProtB  GeneA  GeneB\n",
       "56    O00141   P49815   SGK1   TSC2\n",
       "66    O00141   O00141   SGK1   SGK1\n",
       "160   O43143   P01106  DHX15    MYC\n",
       "294   O43143   P06748  DHX15   NPM1\n",
       "384   O60674   P78347   JAK2  GTF2I"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of interactions amongst the seed genes (Among blue nodes and blue line in the diagram)\n",
    "df_iid_seed_genes = df_iid.loc[df_iid['GeneA'].isin(seed_genes_iid) & df_iid['GeneB'].isin(seed_genes_iid)]\n",
    "print(\"The Number of interactions between the seed genes\",len(df_iid_seed_genes))\n",
    "df_iid_seed_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'>** Interaction among Seed and Non-seed Genes in IID**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Number of interactions between the seed genes and Non seed genes 8764\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P49411</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>TUFM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q14697</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>GANAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q99759</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAP3K3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P10636</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P04406</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>GAPDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniProtA UniProtB GeneA   GeneB\n",
       "0   O00141   P49411  SGK1    TUFM\n",
       "1   O00141   Q14697  SGK1   GANAB\n",
       "2   O00141   Q99759  SGK1  MAP3K3\n",
       "3   O00141   P10636  SGK1    MAPT\n",
       "4   O00141   P04406  SGK1   GAPDH"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of interactions between the seed genes and non seed genes (Between blue and yellow nodes with blue line in the diagram)\n",
    "df_iid_seed_non_seed = df_iid.loc[df_iid['GeneA'].isin(seed_genes_iid) | df_iid['GeneB'].isin(seed_genes_iid)]\n",
    "print(\"The Number of interactions between the seed genes and Non seed genes\",len(df_iid_seed_non_seed))\n",
    "df_iid_seed_non_seed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'>**Getting count of Non-seed genes in IID**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of seed genes: 106\n",
      "\n",
      "Number of non seed genes present in IID are  4653\n",
      "\n",
      "Total no. of interacting genes (seed genes and non seed genes): 4759\n"
     ]
    }
   ],
   "source": [
    "print('The number of seed genes: {}'.format(len(seed_genes_iid)))\n",
    "print('')\n",
    "\n",
    "df_iid_seed = df_iid.loc[df_iid['GeneA'].isin(seed_genes_iid) | df_iid['GeneB'].isin(seed_genes_iid)]\n",
    "\n",
    "#Number of non seed genes (red dotted line in the diagram)\n",
    "non_seed_genesA = set(df_iid_seed['GeneA']).difference(set(seed_genes))\n",
    "non_seed_genesB = set(df_iid_seed['GeneB']).difference(set(seed_genes))\n",
    "\n",
    "non_seed_genes_iid  = list(non_seed_genesA.union(non_seed_genesB))\n",
    "print(\"Number of non seed genes present in IID are \",len(non_seed_genes_iid))\n",
    "print('')\n",
    "all_genes = non_seed_genes_iid + seed_genes_iid\n",
    "print('Total no. of interacting genes (seed genes and non seed genes): {}'.format(len(all_genes)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'>**Non-seed Interactions in IID**</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total Interactions found in IID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total no. of interactions found: 8764\n"
     ]
    }
   ],
   "source": [
    "df_iid_total_interactions = df_iid.loc[df_iid['GeneA'].isin(seed_genes_biogrid) |  df_iid['GeneB'].isin(seed_genes_biogrid) \n",
    "                    | (df_iid['GeneA'].isin(non_seed_genes_iid) & df_iid['GeneB'].isin(non_seed_genes_iid))]\n",
    "\n",
    "print('Total no. of interactions found: {}'.format(len(df_iid)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total no. of interactions found: 8508\n"
     ]
    }
   ],
   "source": [
    "df_iid_final = df_iid.loc[df_iid['GeneA'].isin(seed_genes_biogrid) |  df_iid['GeneB'].isin(seed_genes_biogrid) \n",
    "                    | (df_iid['GeneA'].isin(non_seed_genes_iid) & df_iid['GeneB'].isin(non_seed_genes_iid))]\n",
    "\n",
    "print('Total no. of interactions found: {}'.format(len(df_iid_final)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P49411</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>TUFM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q14697</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>GANAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O00141</td>\n",
       "      <td>Q99759</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAP3K3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P10636</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O00141</td>\n",
       "      <td>P04406</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>GAPDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniProtA UniProtB GeneA   GeneB\n",
       "0   O00141   P49411  SGK1    TUFM\n",
       "1   O00141   Q14697  SGK1   GANAB\n",
       "2   O00141   Q99759  SGK1  MAP3K3\n",
       "3   O00141   P10636  SGK1    MAPT\n",
       "4   O00141   P04406  SGK1   GAPDH"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Storing all the IID interactions data into an easily accessible excel sheet format.\n",
    "\n",
    "writer_object = pd.ExcelWriter('master_df_iid.xlsx',engine='xlsxwriter')\n",
    "\n",
    "df_iid.to_excel(writer_object,sheet_name='df_iid',encoding='utf8',index=False)\n",
    "df_iid_seed_genes.to_excel(writer_object,sheet_name='df_iid_seed_genes',encoding='utf8',index=False)\n",
    "df_iid_seed_non_seed.to_excel(writer_object,sheet_name='df_iid_seed_non_seed',encoding='utf8',index=False)\n",
    "df_iid_total_interactions.to_excel(writer_object,sheet_name='df_iid_non_seed',encoding='utf8',index=False)\n",
    "\n",
    "writer_object.save()\n",
    "\n",
    "df_iid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 ARRANGING DATA INTO INTERACTOMES\n",
    "\n",
    "- <font color='blue'>** A) Seed Genes Interactome**</font>\n",
    "\n",
    "**interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B Uniprot AC, database source **\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Build and store three tables:\n",
    "#### First we create map between GeneSymbol and UniProtAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of human-genes found in uniprot:  188349\n",
      "The number of human-genes having uniprotAC:  26682\n"
     ]
    }
   ],
   "source": [
    "uniprot_all = pd.read_csv('1.3. uniprot-filtered-organism__Homo+sapiens+(Human)+[9606]_.tab', sep='\\t')\n",
    "uniprot_all.rename(columns={ 'Gene names  (primary )':'GeneSymbol', 'Entry': 'UniProtAC'}, inplace=True)\n",
    "print('The number of human-genes found in uniprot: ', len(uniprot_all))\n",
    "uniprot_all.head(3)\n",
    "\n",
    "gene_uniprot = {}\n",
    "count = 0\n",
    "for i in range(len(uniprot_all)):\n",
    "    contents = uniprot_all['GeneSymbol'].iloc[i]\n",
    "    \n",
    "    if not isinstance(contents, float):\n",
    "        contents = contents.split('; ')\n",
    "        for content in contents:\n",
    "            if content not in gene_uniprot.keys():\n",
    "                gene_uniprot[content] = uniprot_all['UniProtAC'].iloc[i]\n",
    "    \n",
    "print('The number of human-genes having uniprotAC: ', len(gene_uniprot))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of seed-seed interactions in BioGrid : 182\n",
      "Number of seed-seed interactions in IID : 162\n",
      "Number of seed genes interactome (seed-seed interactions) in both DB : 344\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Database</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>Q03164</td>\n",
       "      <td>Q03164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>ASMTL</td>\n",
       "      <td>ASMTL</td>\n",
       "      <td>O95671</td>\n",
       "      <td>O95671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>MYC</td>\n",
       "      <td>Q9Y6K1</td>\n",
       "      <td>P01106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>MYC</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>P01106</td>\n",
       "      <td>Q9Y6K1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>CTNNA1</td>\n",
       "      <td>CTNNA1</td>\n",
       "      <td>P35221</td>\n",
       "      <td>P35221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Database   GeneA   GeneB UniProtA UniProtB\n",
       "0  BioGrid   KMT2A   KMT2A   Q03164   Q03164\n",
       "1  BioGrid   ASMTL   ASMTL   O95671   O95671\n",
       "2  BioGrid  DNMT3A     MYC   Q9Y6K1   P01106\n",
       "3  BioGrid     MYC  DNMT3A   P01106   Q9Y6K1\n",
       "4  BioGrid  CTNNA1  CTNNA1   P35221   P35221"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BioGrid\n",
    "# extract only seed-seed interactions\n",
    "#df_seedseed_a_biogrid = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) & df_biogrid['GeneB'].isin(seed_genes_biogrid)]\n",
    "df_seedseed_a_biogrid = df_biogrid_final.loc[df_biogrid_final['GeneA'].isin(seed_genes_biogrid) & df_biogrid_final['GeneB'].isin(seed_genes_biogrid)]\n",
    "df_seedseed_a_biogrid = df_seedseed_a_biogrid.reset_index(drop=True)\n",
    "\n",
    "UniProtA, UniProtB = [], []\n",
    "for i in range(len(df_seedseed_a_biogrid)):\n",
    "    UniProtA.append(gene_uniprot[df_seedseed_a_biogrid['GeneA'].iloc[i]])\n",
    "    UniProtB.append(gene_uniprot[df_seedseed_a_biogrid['GeneB'].iloc[i]])\n",
    "\n",
    "df_seedseed_a_biogrid['UniProtA'] = UniProtA\n",
    "df_seedseed_a_biogrid['UniProtB'] = UniProtB\n",
    "df_seedseed_a_biogrid['Database'] = 'BioGrid'\n",
    "\n",
    "df_seedseed_a_biogrid = df_seedseed_a_biogrid.drop_duplicates(subset=['GeneA', 'GeneB'])\n",
    "\n",
    "print('Number of seed-seed interactions in BioGrid : {}'.format(len(df_seedseed_a_biogrid)))\n",
    "df_seedseed_a_biogrid.head()\n",
    "\n",
    "#IID \n",
    "\n",
    "# extract only seed-seed interactions\n",
    "df_seedseed_a_iid = df_iid.loc[df_iid['GeneA'].isin(seed_genes_iid) & df_iid['GeneB'].isin(seed_genes_iid)]\n",
    "\n",
    "df_seedseed_a_iid['Database'] = 'IID'\n",
    "\n",
    "df_seedseed_a_iid = df_seedseed_a_iid.drop_duplicates(subset=['GeneA','GeneB'])\n",
    "\n",
    "print('Number of seed-seed interactions in IID : {}'.format(len(df_seedseed_a_iid)))\n",
    "df_seedseed_a_iid.head()\n",
    "\n",
    "# concating both data frames\n",
    "# All ------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "df_seedseed_a = pd.concat([df_seedseed_a_biogrid, df_seedseed_a_iid])\n",
    "df_seedseed_a = df_seedseed_a.reset_index(drop=True)\n",
    "\n",
    "df_seedseed_a.to_csv('1.3_a_seed_seed_interactome_table.csv',encoding='utf8',index=False)\n",
    "\n",
    "print('Number of seed genes interactome (seed-seed interactions) in both DB : {}'.format(len(df_seedseed_a)))\n",
    "df_seedseed_a.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- <font color='blue'>** B) Union Interactome**</font>\n",
    "\n",
    "All proteins interacting with at least one seed gene,from all DBs, same format as above.\n",
    "\n",
    "**interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B Uniprot AC, database source **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes that do not exist in UniProt:9487\n",
      "Number of interactions in BioGrid : 198286 \n",
      "Number of interactions in IID : 8748\n",
      "Number of interactions in union interactome in both DB : 207034\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Database</th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>P23769</td>\n",
       "      <td>P29590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>P15927</td>\n",
       "      <td>P40763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BioGrid</td>\n",
       "      <td>ARF1</td>\n",
       "      <td>GGA3</td>\n",
       "      <td>P84077</td>\n",
       "      <td>Q9NZ52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Database  GeneA  GeneB UniProtA UniProtB\n",
       "0  BioGrid  GATA2    PML   P23769   P29590\n",
       "1  BioGrid   RPA2  STAT3   P15927   P40763\n",
       "2  BioGrid   ARF1   GGA3   P84077   Q9NZ52"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# biogrid --------------------------------------------------------------------------------------------------------------------------\n",
    " # extract interactions with at least one seed gene\n",
    "#df_union_b_biogrid = df_biogrid.loc[df_biogrid['GeneA'].isin(seed_genes_biogrid) | df_biogrid['GeneB'].isin(seed_genes_biogrid)]\n",
    "df_union_b_biogrid = df_biogrid_final.copy()\n",
    "df_union_b_biogrid = df_union_b_biogrid.reset_index(drop=True)\n",
    "\n",
    "UniProtA, UniProtB = [], []\n",
    "non_approved_genes = []\n",
    "for i in range(len(df_union_b_biogrid)):\n",
    "    geneA = df_union_b_biogrid['GeneA'].iloc[i]\n",
    "    geneB= df_union_b_biogrid['GeneB'].iloc[i]\n",
    "    try:\n",
    "        uniprotA = gene_uniprot[geneA]\n",
    "        uniprotB = gene_uniprot[geneB]\n",
    "        \n",
    "        UniProtA.append(uniprotA)\n",
    "        UniProtB.append(uniprotB)\n",
    "        \n",
    "    except:\n",
    "        non_approved_genes.append(i)\n",
    "\n",
    "# dropping the non_approved genes        \n",
    "df_union_b_biogrid = df_union_b_biogrid.drop(non_approved_genes)\n",
    "\n",
    "df_union_b_biogrid['UniProtA'] = UniProtA\n",
    "df_union_b_biogrid['UniProtB'] = UniProtB\n",
    "df_union_b_biogrid['Database'] = 'BioGrid'\n",
    "\n",
    "df_union_b_biogrid = df_union_b_biogrid.drop_duplicates(subset=['GeneA', 'GeneB'])\n",
    "\n",
    "# print('Some genes are removed since they do not exist on UniProt')\n",
    "print('Number of genes that do not exist in UniProt:{}'.format(len(non_approved_genes)))\n",
    "print('Number of interactions in BioGrid : {} '.format(len(df_union_b_biogrid)))\n",
    "\n",
    "# iid --------------------------------------------------------------------------------------------------------------------------\n",
    " # extract interactions with at least one seed gene\n",
    "df_union_b_iid = df_iid.loc[df_iid['GeneA'].isin(seed_genes_iid) | df_iid['GeneB'].isin(seed_genes_iid)]\n",
    "df_union_b_iid['Database'] = 'IID'\n",
    "\n",
    "df_union_b_iid = df_union_b_iid.drop_duplicates(subset=['GeneA', 'GeneB'])\n",
    "\n",
    "print('Number of interactions in IID : {}'.format(len(df_union_b_iid)))\n",
    "\n",
    "# All ------------------------------------------------------------------------------------------------------------------------\n",
    "df_union_b = pd.concat([df_union_b_biogrid, df_union_b_iid])\n",
    "df_union_b = df_union_b.reset_index(drop=True)\n",
    "\n",
    "df_union_b.to_csv('1.3_b_union_interactome_table.csv',encoding='utf8',index=False)\n",
    "\n",
    "print('Number of interactions in union interactome in both DB : {}'.format(len(df_union_b)))\n",
    "df_union_b.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- <font color='blue'>** C) Intersection Interactome**</font>\n",
    "\n",
    "All proteins interacting with at least one seed gene confirmed\n",
    "\n",
    "**interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B Uniprot AC, database source **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of intersection interactome in both DB : 3943\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneA</th>\n",
       "      <th>GeneB</th>\n",
       "      <th>UniProtA</th>\n",
       "      <th>UniProtB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGK1</td>\n",
       "      <td>TUFM</td>\n",
       "      <td>O00141</td>\n",
       "      <td>P49411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SGK1</td>\n",
       "      <td>MAPT</td>\n",
       "      <td>O00141</td>\n",
       "      <td>P10636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGK1</td>\n",
       "      <td>CDKN1B</td>\n",
       "      <td>O00141</td>\n",
       "      <td>P46527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GeneA   GeneB UniProtA UniProtB\n",
       "0  SGK1    TUFM   O00141   P49411\n",
       "1  SGK1    MAPT   O00141   P10636\n",
       "2  SGK1  CDKN1B   O00141   P46527"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Obtain intersection (drop duplicate rows in terms of feartures['GeneA', 'GeneB'])\n",
    "# It went well because I already dropped duplicate rows before concatenating [df_intersection_b_biogrid & df_intersection_b_iid ]\n",
    "df_intersection_c = df_union_b[df_union_b.duplicated(['GeneA', 'GeneB'])]\n",
    "df_intersection_c = df_intersection_c.drop('Database', axis = 1)\n",
    "df_intersection_c = df_intersection_c.reset_index(drop=True)\n",
    "\n",
    "df_intersection_c.to_csv('1.3_c_intersection_interactome_table.csv',encoding='utf8',index=False)\n",
    "\n",
    "print('Number of intersection interactome in both DB : {}'.format(len(df_intersection_c)))\n",
    "df_intersection_c.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Storing all the Interactomes to an easily accessible excel sheet format.\n",
    "\n",
    "writer_object = pd.ExcelWriter('master_df_interactomes.xlsx',engine='xlsxwriter')\n",
    "\n",
    "df_seedseed_a.to_excel(writer_object,sheet_name='df_seedseed_interactome_a',encoding='utf8',index=False)\n",
    "df_union_b.to_excel(writer_object,sheet_name='df_union_interactome_b',encoding='utf8',index=False)\n",
    "df_intersection_c.to_excel(writer_object,sheet_name='df_intersection_interactome_c',encoding='utf8',index=False)\n",
    "\n",
    "writer_object.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.4 Enrichment analysis\n",
    "Using the service [Enrichr](https://amp.pharm.mssm.edu/Enrichr/), find, report in tables and save related charts (8 charts in total) of the overrepresented GO categories (limit to the first 10 for each main category, BP, MF, CL) and the the overrepresented pathways (KEGG 2019 Human) for:\n",
    "\n",
    "- a) the seed genes\n",
    "- b) the union interactome genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a) Seed Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of interactions in seed interactome in both DB : 256\n",
      "Number of seed genes found in both DB : 83\n"
     ]
    }
   ],
   "source": [
    "df_seedseed_a = df_seedseed_a.drop_duplicates(['GeneA', 'GeneB'])\n",
    "df_seedseed_a = df_seedseed_a.drop('Database', axis = 1)\n",
    "df_seedseed_a = df_seedseed_a.reset_index(drop=True)\n",
    "print('Number of interactions in seed interactome in both DB : {}'.format(len(df_seedseed_a)))\n",
    "\n",
    "geneA = list(df_seedseed_a['GeneA'].values)\n",
    "geneB = list(df_seedseed_a['GeneB'].values)\n",
    "genes = list(set(geneA + geneB))\n",
    "genes = pd.Series(genes)\n",
    "genes.to_csv('1.4. seed_genes.csv', index = False)\n",
    "print('Number of seed genes found in both DB : {}'.format(len(genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\1.4 EnrichmentAnalysis\\GO_Analysis\\seed_seed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>regulation of apoptotic process (GO:0042981)</td>\n",
       "      <td>23/815</td>\n",
       "      <td>1.435630e-13</td>\n",
       "      <td>7.326018e-10</td>\n",
       "      <td>6.800207</td>\n",
       "      <td>201.095739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>negative regulation of programmed cell death (...</td>\n",
       "      <td>17/408</td>\n",
       "      <td>7.843325e-13</td>\n",
       "      <td>2.001224e-09</td>\n",
       "      <td>10.040161</td>\n",
       "      <td>279.858870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>regulation of myeloid cell differentiation (GO...</td>\n",
       "      <td>9/65</td>\n",
       "      <td>6.186446e-12</td>\n",
       "      <td>1.052314e-08</td>\n",
       "      <td>33.364226</td>\n",
       "      <td>861.085980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>negative regulation of apoptotic process (GO:0...</td>\n",
       "      <td>17/485</td>\n",
       "      <td>1.225554e-11</td>\n",
       "      <td>1.563500e-08</td>\n",
       "      <td>8.446156</td>\n",
       "      <td>212.210029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cytokine-mediated signaling pathway (GO:0019221)</td>\n",
       "      <td>18/633</td>\n",
       "      <td>8.774059e-11</td>\n",
       "      <td>8.954805e-08</td>\n",
       "      <td>6.852053</td>\n",
       "      <td>158.670495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>negative regulation of cell proliferation (GO:...</td>\n",
       "      <td>13/363</td>\n",
       "      <td>3.150347e-09</td>\n",
       "      <td>2.679370e-06</td>\n",
       "      <td>8.629560</td>\n",
       "      <td>168.930129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cellular response to cytokine stimulus (GO:007...</td>\n",
       "      <td>14/456</td>\n",
       "      <td>5.296174e-09</td>\n",
       "      <td>3.860911e-06</td>\n",
       "      <td>7.398013</td>\n",
       "      <td>140.978617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>regulation of cell proliferation (GO:0042127)</td>\n",
       "      <td>17/740</td>\n",
       "      <td>7.885399e-09</td>\n",
       "      <td>5.029899e-06</td>\n",
       "      <td>5.535656</td>\n",
       "      <td>103.285673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>positive regulation of transcription, DNA-temp...</td>\n",
       "      <td>20/1120</td>\n",
       "      <td>2.173566e-08</td>\n",
       "      <td>1.232412e-05</td>\n",
       "      <td>4.302926</td>\n",
       "      <td>75.922167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>positive regulation of peptidyl-tyrosine phosp...</td>\n",
       "      <td>8/116</td>\n",
       "      <td>2.754785e-08</td>\n",
       "      <td>1.405767e-05</td>\n",
       "      <td>16.618197</td>\n",
       "      <td>289.278629</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Term  Overlap       P-value  \\\n",
       "0       regulation of apoptotic process (GO:0042981)   23/815  1.435630e-13   \n",
       "1  negative regulation of programmed cell death (...   17/408  7.843325e-13   \n",
       "2  regulation of myeloid cell differentiation (GO...     9/65  6.186446e-12   \n",
       "3  negative regulation of apoptotic process (GO:0...   17/485  1.225554e-11   \n",
       "4   cytokine-mediated signaling pathway (GO:0019221)   18/633  8.774059e-11   \n",
       "5  negative regulation of cell proliferation (GO:...   13/363  3.150347e-09   \n",
       "6  cellular response to cytokine stimulus (GO:007...   14/456  5.296174e-09   \n",
       "7      regulation of cell proliferation (GO:0042127)   17/740  7.885399e-09   \n",
       "8  positive regulation of transcription, DNA-temp...  20/1120  2.173566e-08   \n",
       "9  positive regulation of peptidyl-tyrosine phosp...    8/116  2.754785e-08   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      7.326018e-10    6.800207      201.095739  \n",
       "1      2.001224e-09   10.040161      279.858870  \n",
       "2      1.052314e-08   33.364226      861.085980  \n",
       "3      1.563500e-08    8.446156      212.210029  \n",
       "4      8.954805e-08    6.852053      158.670495  \n",
       "5      2.679370e-06    8.629560      168.930129  \n",
       "6      3.860911e-06    7.398013      140.978617  \n",
       "7      5.029899e-06    5.535656      103.285673  \n",
       "8      1.232412e-05    4.302926       75.922167  \n",
       "9      1.405767e-05   16.618197      289.278629  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GO Biological Process 2018\n",
    "seed_enricher_BP = pd.read_csv(os.path.join(path , 'GO_Biological_Process_2018_table.txt'),sep='\\t')\n",
    "seed_enricher_BP = seed_enricher_BP[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "seed_enricher_BP.to_csv('1.4_a_GO_Biological_Process_2018_table.csv',encoding='utf8',index=False)\n",
    "seed_enricher_BP.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C2H2 zinc finger domain binding (GO:0070742)</td>\n",
       "      <td>3/12</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.016986</td>\n",
       "      <td>60.240964</td>\n",
       "      <td>670.106201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein homodimerization activity (GO:0042803)</td>\n",
       "      <td>12/664</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.010480</td>\n",
       "      <td>4.354768</td>\n",
       "      <td>47.526086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>transcription regulatory region DNA binding (G...</td>\n",
       "      <td>9/374</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.009465</td>\n",
       "      <td>5.798595</td>\n",
       "      <td>61.522855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein tyrosine kinase activity (GO:0004713)</td>\n",
       "      <td>6/147</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.009704</td>\n",
       "      <td>9.835259</td>\n",
       "      <td>101.276900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transcriptional activator activity, RNA polyme...</td>\n",
       "      <td>7/284</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.041146</td>\n",
       "      <td>5.939250</td>\n",
       "      <td>51.253177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>protein kinase activity (GO:0004672)</td>\n",
       "      <td>9/513</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.052452</td>\n",
       "      <td>4.227436</td>\n",
       "      <td>34.683915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>regulatory region DNA binding (GO:0000975)</td>\n",
       "      <td>6/224</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>0.055858</td>\n",
       "      <td>6.454389</td>\n",
       "      <td>51.553894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>phospholipase inhibitor activity (GO:0004859)</td>\n",
       "      <td>2/8</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.067444</td>\n",
       "      <td>60.240964</td>\n",
       "      <td>461.771114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>histone-lysine N-methyltransferase activity (G...</td>\n",
       "      <td>3/41</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>0.083847</td>\n",
       "      <td>17.631502</td>\n",
       "      <td>129.237517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>core promoter binding (GO:0001047)</td>\n",
       "      <td>4/97</td>\n",
       "      <td>0.000713</td>\n",
       "      <td>0.082020</td>\n",
       "      <td>9.936654</td>\n",
       "      <td>72.006898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Term Overlap   P-value  \\\n",
       "0       C2H2 zinc finger domain binding (GO:0070742)    3/12  0.000015   \n",
       "1     protein homodimerization activity (GO:0042803)  12/664  0.000018   \n",
       "2  transcription regulatory region DNA binding (G...   9/374  0.000025   \n",
       "3      protein tyrosine kinase activity (GO:0004713)   6/147  0.000034   \n",
       "4  transcriptional activator activity, RNA polyme...   7/284  0.000179   \n",
       "5               protein kinase activity (GO:0004672)   9/513  0.000273   \n",
       "6         regulatory region DNA binding (GO:0000975)   6/224  0.000340   \n",
       "7      phospholipase inhibitor activity (GO:0004859)     2/8  0.000469   \n",
       "8  histone-lysine N-methyltransferase activity (G...    3/41  0.000656   \n",
       "9                 core promoter binding (GO:0001047)    4/97  0.000713   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0          0.016986   60.240964      670.106201  \n",
       "1          0.010480    4.354768       47.526086  \n",
       "2          0.009465    5.798595       61.522855  \n",
       "3          0.009704    9.835259      101.276900  \n",
       "4          0.041146    5.939250       51.253177  \n",
       "5          0.052452    4.227436       34.683915  \n",
       "6          0.055858    6.454389       51.553894  \n",
       "7          0.067444   60.240964      461.771114  \n",
       "8          0.083847   17.631502      129.237517  \n",
       "9          0.082020    9.936654       72.006898  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GO_Molecular_Function_2018\n",
    "seed_enricher_MF = pd.read_csv(os.path.join(path , 'GO_Molecular_Function_2018_table.txt'),sep='\\t')\n",
    "seed_enricher_MF = seed_enricher_MF[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "seed_enricher_MF.to_csv('1.4_a_GO_Molecular_Function_2018_table.csv',encoding='utf8',index=False)\n",
    "seed_enricher_MF.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>focal adhesion (GO:0005925)</td>\n",
       "      <td>15/356</td>\n",
       "      <td>1.729093e-11</td>\n",
       "      <td>7.711753e-09</td>\n",
       "      <td>10.152971</td>\n",
       "      <td>251.599153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromatin (GO:0000785)</td>\n",
       "      <td>9/296</td>\n",
       "      <td>3.802348e-06</td>\n",
       "      <td>8.479236e-04</td>\n",
       "      <td>7.326604</td>\n",
       "      <td>91.435221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nuclear periphery (GO:0034399)</td>\n",
       "      <td>5/78</td>\n",
       "      <td>1.813888e-05</td>\n",
       "      <td>2.696647e-03</td>\n",
       "      <td>15.446401</td>\n",
       "      <td>168.635352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nuclear matrix (GO:0016363)</td>\n",
       "      <td>4/59</td>\n",
       "      <td>1.054787e-04</td>\n",
       "      <td>1.176088e-02</td>\n",
       "      <td>16.336533</td>\n",
       "      <td>149.593650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>membrane raft (GO:0045121)</td>\n",
       "      <td>5/119</td>\n",
       "      <td>1.374077e-04</td>\n",
       "      <td>1.225676e-02</td>\n",
       "      <td>10.124532</td>\n",
       "      <td>90.032989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nuclear chromatin (GO:0000790)</td>\n",
       "      <td>6/253</td>\n",
       "      <td>6.462466e-04</td>\n",
       "      <td>4.803766e-02</td>\n",
       "      <td>5.714558</td>\n",
       "      <td>41.969595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>secretory granule lumen (GO:0034774)</td>\n",
       "      <td>6/317</td>\n",
       "      <td>2.052377e-03</td>\n",
       "      <td>1.307657e-01</td>\n",
       "      <td>4.560830</td>\n",
       "      <td>28.225868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>platelet alpha granule membrane (GO:0031092)</td>\n",
       "      <td>2/17</td>\n",
       "      <td>2.222324e-03</td>\n",
       "      <td>1.238946e-01</td>\n",
       "      <td>28.348689</td>\n",
       "      <td>173.187863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tertiary granule (GO:0070820)</td>\n",
       "      <td>4/164</td>\n",
       "      <td>4.845662e-03</td>\n",
       "      <td>2.401295e-01</td>\n",
       "      <td>5.877167</td>\n",
       "      <td>31.323370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nuclear chromosome part (GO:0044454)</td>\n",
       "      <td>6/392</td>\n",
       "      <td>5.797451e-03</td>\n",
       "      <td>2.585663e-01</td>\n",
       "      <td>3.688222</td>\n",
       "      <td>18.995587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Term Overlap       P-value  \\\n",
       "0                   focal adhesion (GO:0005925)  15/356  1.729093e-11   \n",
       "1                        chromatin (GO:0000785)   9/296  3.802348e-06   \n",
       "2                nuclear periphery (GO:0034399)    5/78  1.813888e-05   \n",
       "3                   nuclear matrix (GO:0016363)    4/59  1.054787e-04   \n",
       "4                    membrane raft (GO:0045121)   5/119  1.374077e-04   \n",
       "5                nuclear chromatin (GO:0000790)   6/253  6.462466e-04   \n",
       "6          secretory granule lumen (GO:0034774)   6/317  2.052377e-03   \n",
       "7  platelet alpha granule membrane (GO:0031092)    2/17  2.222324e-03   \n",
       "8                 tertiary granule (GO:0070820)   4/164  4.845662e-03   \n",
       "9          nuclear chromosome part (GO:0044454)   6/392  5.797451e-03   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      7.711753e-09   10.152971      251.599153  \n",
       "1      8.479236e-04    7.326604       91.435221  \n",
       "2      2.696647e-03   15.446401      168.635352  \n",
       "3      1.176088e-02   16.336533      149.593650  \n",
       "4      1.225676e-02   10.124532       90.032989  \n",
       "5      4.803766e-02    5.714558       41.969595  \n",
       "6      1.307657e-01    4.560830       28.225868  \n",
       "7      1.238946e-01   28.348689      173.187863  \n",
       "8      2.401295e-01    5.877167       31.323370  \n",
       "9      2.585663e-01    3.688222       18.995587  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GO_Cellular_Component_2018\n",
    "seed_enricher_CC = pd.read_csv(os.path.join(path , 'GO_Cellular_Component_2018_table.txt'),sep='\\t')\n",
    "seed_enricher_CC = seed_enricher_CC[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "seed_enricher_CC.to_csv('1.4_a_GO_Cellular_Component_2018_table.csv',encoding='utf8',index=False)\n",
    "seed_enricher_CC.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathways in cancer</td>\n",
       "      <td>23/530</td>\n",
       "      <td>1.416777e-17</td>\n",
       "      <td>4.363672e-15</td>\n",
       "      <td>10.456922</td>\n",
       "      <td>405.682169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acute myeloid leukemia</td>\n",
       "      <td>12/66</td>\n",
       "      <td>4.681260e-17</td>\n",
       "      <td>7.209141e-15</td>\n",
       "      <td>43.811610</td>\n",
       "      <td>1647.333154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcriptional misregulation in cancer</td>\n",
       "      <td>16/186</td>\n",
       "      <td>4.888894e-17</td>\n",
       "      <td>5.019265e-15</td>\n",
       "      <td>20.728074</td>\n",
       "      <td>778.483854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PI3K-Akt signaling pathway</td>\n",
       "      <td>15/354</td>\n",
       "      <td>1.596539e-11</td>\n",
       "      <td>1.229335e-09</td>\n",
       "      <td>10.210333</td>\n",
       "      <td>253.834979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>8/143</td>\n",
       "      <td>1.407830e-07</td>\n",
       "      <td>8.672232e-06</td>\n",
       "      <td>13.480495</td>\n",
       "      <td>212.668919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteoglycans in cancer</td>\n",
       "      <td>9/201</td>\n",
       "      <td>1.512087e-07</td>\n",
       "      <td>7.762045e-06</td>\n",
       "      <td>10.789426</td>\n",
       "      <td>169.443680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hematopoietic cell lineage</td>\n",
       "      <td>7/97</td>\n",
       "      <td>1.557689e-07</td>\n",
       "      <td>6.853833e-06</td>\n",
       "      <td>17.389144</td>\n",
       "      <td>272.572961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Central carbon metabolism in cancer</td>\n",
       "      <td>6/65</td>\n",
       "      <td>2.888038e-07</td>\n",
       "      <td>1.111895e-05</td>\n",
       "      <td>22.242817</td>\n",
       "      <td>334.921628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cellular senescence</td>\n",
       "      <td>8/160</td>\n",
       "      <td>3.338168e-07</td>\n",
       "      <td>1.142395e-05</td>\n",
       "      <td>12.048193</td>\n",
       "      <td>179.670765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>JAK-STAT signaling pathway</td>\n",
       "      <td>8/162</td>\n",
       "      <td>3.670589e-07</td>\n",
       "      <td>1.130542e-05</td>\n",
       "      <td>11.899450</td>\n",
       "      <td>176.322991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Term Overlap       P-value  \\\n",
       "0                       Pathways in cancer  23/530  1.416777e-17   \n",
       "1                   Acute myeloid leukemia   12/66  4.681260e-17   \n",
       "2  Transcriptional misregulation in cancer  16/186  4.888894e-17   \n",
       "3               PI3K-Akt signaling pathway  15/354  1.596539e-11   \n",
       "4                                Apoptosis   8/143  1.407830e-07   \n",
       "5                  Proteoglycans in cancer   9/201  1.512087e-07   \n",
       "6               Hematopoietic cell lineage    7/97  1.557689e-07   \n",
       "7      Central carbon metabolism in cancer    6/65  2.888038e-07   \n",
       "8                      Cellular senescence   8/160  3.338168e-07   \n",
       "9               JAK-STAT signaling pathway   8/162  3.670589e-07   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      4.363672e-15   10.456922      405.682169  \n",
       "1      7.209141e-15   43.811610     1647.333154  \n",
       "2      5.019265e-15   20.728074      778.483854  \n",
       "3      1.229335e-09   10.210333      253.834979  \n",
       "4      8.672232e-06   13.480495      212.668919  \n",
       "5      7.762045e-06   10.789426      169.443680  \n",
       "6      6.853833e-06   17.389144      272.572961  \n",
       "7      1.111895e-05   22.242817      334.921628  \n",
       "8      1.142395e-05   12.048193      179.670765  \n",
       "9      1.130542e-05   11.899450      176.322991  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KEGG 2019 HUMAN\n",
    "kegg_human_seed_enrichr = pd.read_csv(os.path.join(path,'KEGG_2019_Human_table.txt'),sep='\\t')\n",
    "kegg_human_seed_enrichr = kegg_human_seed_enrichr[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "kegg_human_seed_enrichr.to_csv('1.4_a_KEGG_2019_Human_table.csv',encoding='utf8',index=False)\n",
    "kegg_human_seed_enrichr.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b) Union Interactome Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of interactions in union interactome in both DB : 203091\n",
      "Number of seed genes found in both DB : 6621\n"
     ]
    }
   ],
   "source": [
    "df_union_b = df_union_b.drop_duplicates(subset=['GeneA', 'GeneB'])\n",
    "#df_union_b = df_union_b.drop('Database', axis = 1)\n",
    "df_union_b = df_union_b.reset_index(drop=True)\n",
    "print('Number of interactions in union interactome in both DB : {}'.format(len(df_union_b)))\n",
    "\n",
    "geneA = list(df_union_b['GeneA'].values)\n",
    "geneB = list(df_union_b['GeneB'].values)\n",
    "genes = list(set(geneA + geneB))\n",
    "genes = pd.Series(genes)\n",
    "genes.to_csv('1.4. union_interacome_genes.csv', index = False)\n",
    "print('Number of seed genes found in both DB : {}'.format(len(genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>regulation of transcription from RNA polymeras...</td>\n",
       "      <td>812/1478</td>\n",
       "      <td>4.234328e-72</td>\n",
       "      <td>2.160777e-68</td>\n",
       "      <td>1.659541</td>\n",
       "      <td>272.733795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>positive regulation of gene expression (GO:001...</td>\n",
       "      <td>488/771</td>\n",
       "      <td>1.768360e-68</td>\n",
       "      <td>4.511970e-65</td>\n",
       "      <td>1.911929</td>\n",
       "      <td>298.271949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>positive regulation of transcription, DNA-temp...</td>\n",
       "      <td>644/1120</td>\n",
       "      <td>1.069545e-66</td>\n",
       "      <td>1.819297e-63</td>\n",
       "      <td>1.736898</td>\n",
       "      <td>263.840643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>regulation of apoptotic process (GO:0042981)</td>\n",
       "      <td>498/815</td>\n",
       "      <td>1.085115e-62</td>\n",
       "      <td>1.384336e-59</td>\n",
       "      <td>1.845772</td>\n",
       "      <td>263.352199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>negative regulation of transcription, DNA-temp...</td>\n",
       "      <td>487/813</td>\n",
       "      <td>1.669707e-57</td>\n",
       "      <td>1.704103e-54</td>\n",
       "      <td>1.809443</td>\n",
       "      <td>236.556946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gene expression (GO:0010467)</td>\n",
       "      <td>293/411</td>\n",
       "      <td>1.721155e-57</td>\n",
       "      <td>1.463842e-54</td>\n",
       "      <td>2.153437</td>\n",
       "      <td>281.463617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ribosome biogenesis (GO:0042254)</td>\n",
       "      <td>189/226</td>\n",
       "      <td>6.355211e-56</td>\n",
       "      <td>4.632949e-53</td>\n",
       "      <td>2.526154</td>\n",
       "      <td>321.062744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rRNA metabolic process (GO:0016072)</td>\n",
       "      <td>172/200</td>\n",
       "      <td>1.248564e-54</td>\n",
       "      <td>7.964276e-52</td>\n",
       "      <td>2.597795</td>\n",
       "      <td>322.432071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rRNA processing (GO:0006364)</td>\n",
       "      <td>172/202</td>\n",
       "      <td>2.666323e-53</td>\n",
       "      <td>1.511805e-50</td>\n",
       "      <td>2.572074</td>\n",
       "      <td>311.365805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ncRNA processing (GO:0034470)</td>\n",
       "      <td>186/227</td>\n",
       "      <td>2.405500e-52</td>\n",
       "      <td>1.227527e-49</td>\n",
       "      <td>2.475104</td>\n",
       "      <td>294.182641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Term   Overlap       P-value  \\\n",
       "0  regulation of transcription from RNA polymeras...  812/1478  4.234328e-72   \n",
       "1  positive regulation of gene expression (GO:001...   488/771  1.768360e-68   \n",
       "2  positive regulation of transcription, DNA-temp...  644/1120  1.069545e-66   \n",
       "3       regulation of apoptotic process (GO:0042981)   498/815  1.085115e-62   \n",
       "4  negative regulation of transcription, DNA-temp...   487/813  1.669707e-57   \n",
       "5                       gene expression (GO:0010467)   293/411  1.721155e-57   \n",
       "6                   ribosome biogenesis (GO:0042254)   189/226  6.355211e-56   \n",
       "7                rRNA metabolic process (GO:0016072)   172/200  1.248564e-54   \n",
       "8                       rRNA processing (GO:0006364)   172/202  2.666323e-53   \n",
       "9                      ncRNA processing (GO:0034470)   186/227  2.405500e-52   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      2.160777e-68    1.659541      272.733795  \n",
       "1      4.511970e-65    1.911929      298.271949  \n",
       "2      1.819297e-63    1.736898      263.840643  \n",
       "3      1.384336e-59    1.845772      263.352199  \n",
       "4      1.704103e-54    1.809443      236.556946  \n",
       "5      1.463842e-54    2.153437      281.463617  \n",
       "6      4.632949e-53    2.526154      321.062744  \n",
       "7      7.964276e-52    2.597795      322.432071  \n",
       "8      1.511805e-50    2.572074      311.365805  \n",
       "9      1.227527e-49    2.475104      294.182641  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# enrichment analysis for union interactome\n",
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\1.4 EnrichmentAnalysis\\GO_Analysis\\union_interactome'\n",
    "union_enricher_BP = pd.read_csv(os.path.join(path , 'GO_Biological_Process_2018_table.txt'),sep='\\t')\n",
    "union_enricher_BP = union_enricher_BP[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "union_enricher_BP.to_csv('1.4_b_GO_Biological_Process_2018_table.csv',encoding='utf8',index=False)\n",
    "union_enricher_BP.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNA binding (GO:0003723)</td>\n",
       "      <td>919/1387</td>\n",
       "      <td>2.199402e-151</td>\n",
       "      <td>2.531512e-148</td>\n",
       "      <td>2.001453</td>\n",
       "      <td>694.308475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cadherin binding (GO:0045296)</td>\n",
       "      <td>249/313</td>\n",
       "      <td>7.639852e-65</td>\n",
       "      <td>4.396735e-62</td>\n",
       "      <td>2.403042</td>\n",
       "      <td>354.772317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein kinase binding (GO:0019901)</td>\n",
       "      <td>333/495</td>\n",
       "      <td>1.020128e-55</td>\n",
       "      <td>3.913893e-53</td>\n",
       "      <td>2.032102</td>\n",
       "      <td>257.309295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein kinase activity (GO:0004672)</td>\n",
       "      <td>329/513</td>\n",
       "      <td>5.596061e-48</td>\n",
       "      <td>1.610267e-45</td>\n",
       "      <td>1.937247</td>\n",
       "      <td>210.776362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>kinase binding (GO:0019900)</td>\n",
       "      <td>277/418</td>\n",
       "      <td>1.624994e-44</td>\n",
       "      <td>3.740735e-42</td>\n",
       "      <td>2.001750</td>\n",
       "      <td>201.832957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>transcription regulatory region DNA binding (G...</td>\n",
       "      <td>247/374</td>\n",
       "      <td>2.081648e-39</td>\n",
       "      <td>3.993294e-37</td>\n",
       "      <td>1.994949</td>\n",
       "      <td>177.685421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ubiquitin-like protein ligase binding (GO:0044...</td>\n",
       "      <td>203/297</td>\n",
       "      <td>7.953371e-36</td>\n",
       "      <td>1.307761e-33</td>\n",
       "      <td>2.064648</td>\n",
       "      <td>166.863743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>protein serine/threonine kinase activity (GO:0...</td>\n",
       "      <td>237/368</td>\n",
       "      <td>3.208229e-35</td>\n",
       "      <td>4.615839e-33</td>\n",
       "      <td>1.945391</td>\n",
       "      <td>154.512223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>transcription coactivator activity (GO:0003713)</td>\n",
       "      <td>199/291</td>\n",
       "      <td>3.429370e-35</td>\n",
       "      <td>4.385784e-33</td>\n",
       "      <td>2.065696</td>\n",
       "      <td>163.929746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ubiquitin protein ligase binding (GO:0031625)</td>\n",
       "      <td>191/284</td>\n",
       "      <td>2.569341e-32</td>\n",
       "      <td>2.957311e-30</td>\n",
       "      <td>2.031522</td>\n",
       "      <td>147.770996</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Term   Overlap        P-value  \\\n",
       "0                           RNA binding (GO:0003723)  919/1387  2.199402e-151   \n",
       "1                      cadherin binding (GO:0045296)   249/313   7.639852e-65   \n",
       "2                protein kinase binding (GO:0019901)   333/495   1.020128e-55   \n",
       "3               protein kinase activity (GO:0004672)   329/513   5.596061e-48   \n",
       "4                        kinase binding (GO:0019900)   277/418   1.624994e-44   \n",
       "5  transcription regulatory region DNA binding (G...   247/374   2.081648e-39   \n",
       "6  ubiquitin-like protein ligase binding (GO:0044...   203/297   7.953371e-36   \n",
       "7  protein serine/threonine kinase activity (GO:0...   237/368   3.208229e-35   \n",
       "8    transcription coactivator activity (GO:0003713)   199/291   3.429370e-35   \n",
       "9      ubiquitin protein ligase binding (GO:0031625)   191/284   2.569341e-32   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0     2.531512e-148    2.001453      694.308475  \n",
       "1      4.396735e-62    2.403042      354.772317  \n",
       "2      3.913893e-53    2.032102      257.309295  \n",
       "3      1.610267e-45    1.937247      210.776362  \n",
       "4      3.740735e-42    2.001750      201.832957  \n",
       "5      3.993294e-37    1.994949      177.685421  \n",
       "6      1.307761e-33    2.064648      166.863743  \n",
       "7      4.615839e-33    1.945391      154.512223  \n",
       "8      4.385784e-33    2.065696      163.929746  \n",
       "9      2.957311e-30    2.031522      147.770996  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "union_enricher_MF = pd.read_csv(os.path.join(path , 'GO_Molecular_Function_2018_table.txt'),sep='\\t')\n",
    "union_enricher_MF = union_enricher_MF[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "union_enricher_MF.to_csv('1.4_b_GO_Molecular_Function_2018_table.csv',encoding='utf8',index=False)\n",
    "union_enricher_MF.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>focal adhesion (GO:0005925)</td>\n",
       "      <td>288/356</td>\n",
       "      <td>3.863285e-78</td>\n",
       "      <td>1.723025e-75</td>\n",
       "      <td>2.443706</td>\n",
       "      <td>435.590829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nucleolus (GO:0005730)</td>\n",
       "      <td>449/676</td>\n",
       "      <td>1.618671e-72</td>\n",
       "      <td>3.609635e-70</td>\n",
       "      <td>2.006347</td>\n",
       "      <td>331.658235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nuclear body (GO:0016604)</td>\n",
       "      <td>375/618</td>\n",
       "      <td>4.022621e-46</td>\n",
       "      <td>5.980296e-44</td>\n",
       "      <td>1.832944</td>\n",
       "      <td>191.592104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chromatin (GO:0000785)</td>\n",
       "      <td>203/296</td>\n",
       "      <td>3.721673e-36</td>\n",
       "      <td>4.149666e-34</td>\n",
       "      <td>2.071623</td>\n",
       "      <td>169.000709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nuclear chromosome part (GO:0044454)</td>\n",
       "      <td>243/392</td>\n",
       "      <td>3.231631e-32</td>\n",
       "      <td>2.882615e-30</td>\n",
       "      <td>1.872521</td>\n",
       "      <td>135.775977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nucleoplasm part (GO:0044451)</td>\n",
       "      <td>248/407</td>\n",
       "      <td>3.612529e-31</td>\n",
       "      <td>2.685313e-29</td>\n",
       "      <td>1.840618</td>\n",
       "      <td>129.019466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cytoskeleton (GO:0005856)</td>\n",
       "      <td>297/520</td>\n",
       "      <td>4.705731e-30</td>\n",
       "      <td>2.998223e-28</td>\n",
       "      <td>1.725280</td>\n",
       "      <td>116.506019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cytosolic part (GO:0044445)</td>\n",
       "      <td>121/159</td>\n",
       "      <td>9.044053e-29</td>\n",
       "      <td>5.042059e-27</td>\n",
       "      <td>2.298765</td>\n",
       "      <td>148.437858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nuclear speck (GO:0016607)</td>\n",
       "      <td>187/296</td>\n",
       "      <td>1.618878e-26</td>\n",
       "      <td>8.022442e-25</td>\n",
       "      <td>1.908342</td>\n",
       "      <td>113.327828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>microtubule cytoskeleton (GO:0015630)</td>\n",
       "      <td>229/388</td>\n",
       "      <td>4.205394e-26</td>\n",
       "      <td>1.875606e-24</td>\n",
       "      <td>1.782831</td>\n",
       "      <td>104.172318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Term  Overlap       P-value  \\\n",
       "0            focal adhesion (GO:0005925)  288/356  3.863285e-78   \n",
       "1                 nucleolus (GO:0005730)  449/676  1.618671e-72   \n",
       "2              nuclear body (GO:0016604)  375/618  4.022621e-46   \n",
       "3                 chromatin (GO:0000785)  203/296  3.721673e-36   \n",
       "4   nuclear chromosome part (GO:0044454)  243/392  3.231631e-32   \n",
       "5          nucleoplasm part (GO:0044451)  248/407  3.612529e-31   \n",
       "6              cytoskeleton (GO:0005856)  297/520  4.705731e-30   \n",
       "7            cytosolic part (GO:0044445)  121/159  9.044053e-29   \n",
       "8             nuclear speck (GO:0016607)  187/296  1.618878e-26   \n",
       "9  microtubule cytoskeleton (GO:0015630)  229/388  4.205394e-26   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      1.723025e-75    2.443706      435.590829  \n",
       "1      3.609635e-70    2.006347      331.658235  \n",
       "2      5.980296e-44    1.832944      191.592104  \n",
       "3      4.149666e-34    2.071623      169.000709  \n",
       "4      2.882615e-30    1.872521      135.775977  \n",
       "5      2.685313e-29    1.840618      129.019466  \n",
       "6      2.998223e-28    1.725280      116.506019  \n",
       "7      5.042059e-27    2.298765      148.437858  \n",
       "8      8.022442e-25    1.908342      113.327828  \n",
       "9      1.875606e-24    1.782831      104.172318  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "union_enricher_CC = pd.read_csv(os.path.join(path , 'GO_Cellular_Component_2018_table.txt'),sep='\\t')\n",
    "union_enricher_CC = union_enricher_CC[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "union_enricher_CC.to_csv('1.4_b_GO_Cellular_Component_2018_table.csv',encoding='utf8',index=False)\n",
    "union_enricher_CC.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathways in cancer</td>\n",
       "      <td>332/530</td>\n",
       "      <td>4.042148e-45</td>\n",
       "      <td>1.244981e-42</td>\n",
       "      <td>1.892207</td>\n",
       "      <td>193.420544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human T-cell leukemia virus 1 infection</td>\n",
       "      <td>163/219</td>\n",
       "      <td>3.168471e-36</td>\n",
       "      <td>4.879445e-34</td>\n",
       "      <td>2.248277</td>\n",
       "      <td>183.773778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular senescence</td>\n",
       "      <td>126/160</td>\n",
       "      <td>1.570885e-32</td>\n",
       "      <td>1.612775e-30</td>\n",
       "      <td>2.378795</td>\n",
       "      <td>174.201718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>116/143</td>\n",
       "      <td>2.701171e-32</td>\n",
       "      <td>2.079902e-30</td>\n",
       "      <td>2.450351</td>\n",
       "      <td>178.113680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>104/124</td>\n",
       "      <td>1.597316e-31</td>\n",
       "      <td>9.839468e-30</td>\n",
       "      <td>2.533483</td>\n",
       "      <td>179.653901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Proteoglycans in cancer</td>\n",
       "      <td>143/201</td>\n",
       "      <td>1.849967e-28</td>\n",
       "      <td>9.496499e-27</td>\n",
       "      <td>2.149049</td>\n",
       "      <td>137.232305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Epstein-Barr virus infection</td>\n",
       "      <td>142/201</td>\n",
       "      <td>9.274629e-28</td>\n",
       "      <td>4.080837e-26</td>\n",
       "      <td>2.134021</td>\n",
       "      <td>132.832352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Hepatitis B</td>\n",
       "      <td>121/163</td>\n",
       "      <td>4.770796e-27</td>\n",
       "      <td>1.836757e-25</td>\n",
       "      <td>2.242354</td>\n",
       "      <td>135.902986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Chronic myeloid leukemia</td>\n",
       "      <td>69/76</td>\n",
       "      <td>8.277786e-26</td>\n",
       "      <td>2.832842e-24</td>\n",
       "      <td>2.742470</td>\n",
       "      <td>158.387624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "      <td>105/137</td>\n",
       "      <td>1.386732e-25</td>\n",
       "      <td>4.271134e-24</td>\n",
       "      <td>2.315129</td>\n",
       "      <td>132.512590</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Term  Overlap       P-value  \\\n",
       "0                       Pathways in cancer  332/530  4.042148e-45   \n",
       "1  Human T-cell leukemia virus 1 infection  163/219  3.168471e-36   \n",
       "2                      Cellular senescence  126/160  1.570885e-32   \n",
       "3                                Apoptosis  116/143  2.701171e-32   \n",
       "4                               Cell cycle  104/124  1.597316e-31   \n",
       "5                  Proteoglycans in cancer  143/201  1.849967e-28   \n",
       "6             Epstein-Barr virus infection  142/201  9.274629e-28   \n",
       "7                              Hepatitis B  121/163  4.770796e-27   \n",
       "8                 Chronic myeloid leukemia    69/76  8.277786e-26   \n",
       "9           Ubiquitin mediated proteolysis  105/137  1.386732e-25   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      1.244981e-42    1.892207      193.420544  \n",
       "1      4.879445e-34    2.248277      183.773778  \n",
       "2      1.612775e-30    2.378795      174.201718  \n",
       "3      2.079902e-30    2.450351      178.113680  \n",
       "4      9.839468e-30    2.533483      179.653901  \n",
       "5      9.496499e-27    2.149049      137.232305  \n",
       "6      4.080837e-26    2.134021      132.832352  \n",
       "7      1.836757e-25    2.242354      135.902986  \n",
       "8      2.832842e-24    2.742470      158.387624  \n",
       "9      4.271134e-24    2.315129      132.512590  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KEGG 2019 HUMAN\n",
    "kegg_human_union_enrichr = pd.read_csv(os.path.join(path,'KEGG_2019_Human_table.txt'),sep='\\t')\n",
    "kegg_human_union_enrichr = kegg_human_union_enrichr[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "kegg_human_union_enrichr.to_csv('1.4_b_KEGG_2019_Human_table.csv',encoding='utf8',index=False)\n",
    "kegg_human_union_enrichr.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2 - Data analysis ------------------------------------------------------\n",
    "\n",
    "### Scope of the project:\n",
    "Starting from **the seed genes interactome (SGI), the intersection (I) and the union (U)** interactomes built in the first part of the project, \n",
    "- compute the main network measures for I, U and their nodes, \n",
    "- apply clustering methods for disease modules discovery, \n",
    "- carry on an enrichment analysis on the putative disease modules and \n",
    "- produce a short report.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "256\n",
      "203091\n",
      "3943\n"
     ]
    }
   ],
   "source": [
    "SGI  = df_seedseed_a[['GeneA', 'GeneB']]\n",
    "print(len(SGI))\n",
    "SGI.to_csv('2. SGI.csv', index = False)\n",
    "\n",
    "U = df_union_b[['GeneA', 'GeneB']]\n",
    "print(len(U))\n",
    "U.to_csv('2. U.csv', index = False)\n",
    "\n",
    "I = df_intersection_c[['GeneA', 'GeneB']]\n",
    "print(len(I))\n",
    "I.to_csv('2. I.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1) Calculate the main network measures for SGI, I and U\n",
    "\n",
    "### Calculate the main network measures for SGI, I and U\n",
    "#### a) Calculate the following global (i.e. concerning the whole network and not the single nodes)\n",
    "measures of SGI, U and I (only if no. of nodes >20):\n",
    "\n",
    "- No. of nodes and no. of links\n",
    "-  No. of connected components\n",
    "- No. of isolated nodes\n",
    "- Average path length\n",
    "- Average degree\n",
    "- Average clustering coefficient\n",
    "- Network diameter & radius\n",
    "- Centralization\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Feeding the Genes data from the interactomes into  CytoScape for our Analysis\n",
    "Downloading the required statistics from cytoscape\n",
    "Now we have files named (***2.1. result_SGI.netstats, 2.1. result_I.netstats, 2.1. result_U.netstats***)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactome</th>\n",
       "      <th>nodeCount</th>\n",
       "      <th>nsl</th>\n",
       "      <th>links</th>\n",
       "      <th>ncc</th>\n",
       "      <th>usn</th>\n",
       "      <th>avSpl</th>\n",
       "      <th>cc</th>\n",
       "      <th>avNeighbors</th>\n",
       "      <th>diameter</th>\n",
       "      <th>radius</th>\n",
       "      <th>centralization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGI</td>\n",
       "      <td>83</td>\n",
       "      <td>60</td>\n",
       "      <td>256</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>3.872</td>\n",
       "      <td>0.125</td>\n",
       "      <td>3.566</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>2677</td>\n",
       "      <td>38</td>\n",
       "      <td>3943</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.917</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  interactome nodeCount   nsl   links ncc usn  avSpl     cc avNeighbors  \\\n",
       "0         SGI        83    60     256  14  12  3.872  0.125       3.566   \n",
       "1           I      2677    38    3943   4   0   3.58  0.014       2.917   \n",
       "2           U      6621  1638  203091   1   0  3.073  0.098      55.148   \n",
       "\n",
       "  diameter radius centralization  \n",
       "0        9      1        0.00346  \n",
       "1        9      1          7e-05  \n",
       "2        7      1          5e-05  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loading the files downloaded from cytoscapes\n",
    "stats = ['interactome', 'nodeCount', 'nsl', 'links', 'ncc', 'usn', 'avSpl', 'cc', 'avNeighbors', 'diameter', 'radius','centralization']\n",
    "measure = pd.DataFrame(columns=stats)\n",
    "interactomes = ['SGI', 'I', 'U']\n",
    "\n",
    "for idx, interactome in enumerate(interactomes):\n",
    "    measure.set_value(idx, 'interactome', interactome)\n",
    "    \n",
    "    path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.1_Data_analysis'\n",
    "    filename = os.path.join(path,'2.1. stats_' + interactome + '.netstats')\n",
    "    file = open(filename, \"r\")\n",
    "    \n",
    "    for row in file:\n",
    "        row = row.split(' ')\n",
    "        stat = row[0]\n",
    "\n",
    "        if stat in stats:\n",
    "            value = row[-1].split('\\n')[0]\n",
    "            value = round(float(value), 3)\n",
    "            measure.set_value(idx, stat, value)\n",
    "\n",
    "# We use networkx for centralization\n",
    "import networkx as nx\n",
    "\n",
    "G_SGI = nx.from_pandas_edgelist(SGI, 'GeneA', 'GeneB')\n",
    "G_I = nx.from_pandas_edgelist(I, 'GeneA', 'GeneB')\n",
    "G_U = nx.from_pandas_edgelist(U, 'GeneA', 'GeneB')\n",
    "\n",
    "def centralization(G):\n",
    "    n=G.number_of_nodes()\n",
    "    centralities=nx.degree_centrality(G).values()\n",
    "    Centralization=(n*max(centralities) - sum(centralities))/ (n-1)**2\n",
    "    return round(Centralization, 5)\n",
    "\n",
    "measure.set_value([0,1,2], 'centralization', [centralization(G_SGI), centralization(G_I), centralization(G_U)])\n",
    "\n",
    "measure.set_value([0,1,2], 'links', [len(SGI), len(I), len(U)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global \n",
    "\n",
    "#### b) Isolate the largest connected component (LCC) of I and U and calculate the following global and local (i.e. for each node) measures:\n",
    "i)\n",
    "- N. of nodes and no. of links\n",
    "- Average path length\n",
    "- Average degree\n",
    "- Average clustering coefficient\n",
    "- Network diameter & radius\n",
    "- Centralization\n",
    "\n",
    "#### I-LCC "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3895\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactome</th>\n",
       "      <th>nodeCount</th>\n",
       "      <th>nsl</th>\n",
       "      <th>links</th>\n",
       "      <th>ncc</th>\n",
       "      <th>usn</th>\n",
       "      <th>avSpl</th>\n",
       "      <th>cc</th>\n",
       "      <th>avNeighbors</th>\n",
       "      <th>diameter</th>\n",
       "      <th>radius</th>\n",
       "      <th>centralization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGI</td>\n",
       "      <td>83</td>\n",
       "      <td>60</td>\n",
       "      <td>256</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>3.872</td>\n",
       "      <td>0.125</td>\n",
       "      <td>3.566</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>2677</td>\n",
       "      <td>38</td>\n",
       "      <td>3943</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.917</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>2664</td>\n",
       "      <td>0</td>\n",
       "      <td>3895</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.923</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  interactome nodeCount   nsl   links ncc usn  avSpl     cc avNeighbors  \\\n",
       "0         SGI        83    60     256  14  12  3.872  0.125       3.566   \n",
       "1           I      2677    38    3943   4   0   3.58  0.014       2.917   \n",
       "2           U      6621  1638  203091   1   0  3.073  0.098      55.148   \n",
       "3       I-LCC      2664     0    3895   1   0   3.58  0.014       2.923   \n",
       "\n",
       "  diameter radius centralization  \n",
       "0        9      1        0.00346  \n",
       "1        9      1          7e-05  \n",
       "2        7      1          5e-05  \n",
       "3        9      1          7e-05  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.1_Data_analysis'\n",
    "I_LCC = pd.read_csv(os.path.join(path,'2.1. I_LCC default  edge.csv'))\n",
    "links = len(I_LCC)\n",
    "geneA, geneB = [],[]\n",
    "for i in range(len(I_LCC)):\n",
    "    A = I_LCC.iloc[i]['name'].split(' (interacts with) ')[0]\n",
    "    B = I_LCC.iloc[i]['name'].split(' (interacts with) ')[1]\n",
    "    geneA.append(A)\n",
    "    geneB.append(B)\n",
    "    \n",
    "I_LCC['GeneA'] = geneA\n",
    "I_LCC['GeneB'] = geneB\n",
    "I_LCC = I_LCC[['GeneA', 'GeneB']]\n",
    "print(len(I_LCC))\n",
    "\n",
    "measure.set_value(3, 'interactome', 'I-LCC')\n",
    "filename = os.path.join(path,'2.1. stats_I_LCC.netstats')\n",
    "file = open(filename, \"r\")\n",
    "\n",
    "for row in file:\n",
    "    row = row.split(' ')\n",
    "    stat = row[0]\n",
    "\n",
    "    if stat in stats:\n",
    "        value = row[-1].split('\\n')[0]\n",
    "        value = round(float(value), 3)\n",
    "        measure.set_value(3, stat, value)\n",
    "\n",
    "measure.set_value(3, 'links', links)\n",
    "\n",
    "# centralization -------------------------------------------------\n",
    "G_I_LCC = nx.from_pandas_edgelist(I_LCC, 'GeneA', 'GeneB')\n",
    "\n",
    "measure.set_value(3, 'centralization', centralization(G_I_LCC))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b) Isolate the largest connected component (LCC) of I and U and calculate the following global and local (i.e. for each node) measures:\n",
    "i)\n",
    "- N. of nodes and no. of links\n",
    "- Average path length\n",
    "- Average degree\n",
    "- Average clustering coefficient\n",
    "- Network diameter & radius\n",
    "- Centralization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### U-LCC\n",
    "\n",
    "In our case, U-LCC is the same network as U because U has only one component"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactome</th>\n",
       "      <th>nodeCount</th>\n",
       "      <th>nsl</th>\n",
       "      <th>links</th>\n",
       "      <th>ncc</th>\n",
       "      <th>usn</th>\n",
       "      <th>avSpl</th>\n",
       "      <th>cc</th>\n",
       "      <th>avNeighbors</th>\n",
       "      <th>diameter</th>\n",
       "      <th>radius</th>\n",
       "      <th>centralization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGI</td>\n",
       "      <td>83</td>\n",
       "      <td>60</td>\n",
       "      <td>256</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>3.872</td>\n",
       "      <td>0.125</td>\n",
       "      <td>3.566</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>2677</td>\n",
       "      <td>38</td>\n",
       "      <td>3943</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.917</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>2664</td>\n",
       "      <td>0</td>\n",
       "      <td>3895</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.923</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U-LCC</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  interactome nodeCount   nsl   links ncc usn  avSpl     cc avNeighbors  \\\n",
       "0         SGI        83    60     256  14  12  3.872  0.125       3.566   \n",
       "1           I      2677    38    3943   4   0   3.58  0.014       2.917   \n",
       "2           U      6621  1638  203091   1   0  3.073  0.098      55.148   \n",
       "3       I-LCC      2664     0    3895   1   0   3.58  0.014       2.923   \n",
       "4       U-LCC      6621  1638  203091   1   0  3.073  0.098      55.148   \n",
       "\n",
       "  diameter radius centralization  \n",
       "0        9      1        0.00346  \n",
       "1        9      1          7e-05  \n",
       "2        7      1          5e-05  \n",
       "3        9      1          7e-05  \n",
       "4        7      1          5e-05  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measure = measure.append(measure.iloc[2])\n",
    "measure = measure.reset_index(drop=True)\n",
    "measure.set_value(4, 'interactome', 'U-LCC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactome</th>\n",
       "      <th>nodes</th>\n",
       "      <th>self-loops</th>\n",
       "      <th>links</th>\n",
       "      <th>connected components</th>\n",
       "      <th>isolated nodes</th>\n",
       "      <th>average path length</th>\n",
       "      <th>average clustering coefficient</th>\n",
       "      <th>average degree</th>\n",
       "      <th>diameter</th>\n",
       "      <th>radius</th>\n",
       "      <th>centralization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGI</td>\n",
       "      <td>83</td>\n",
       "      <td>60</td>\n",
       "      <td>256</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>3.872</td>\n",
       "      <td>0.125</td>\n",
       "      <td>3.566</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>2677</td>\n",
       "      <td>38</td>\n",
       "      <td>3943</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.917</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>2664</td>\n",
       "      <td>0</td>\n",
       "      <td>3895</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.58</td>\n",
       "      <td>0.014</td>\n",
       "      <td>2.923</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U-LCC</td>\n",
       "      <td>6621</td>\n",
       "      <td>1638</td>\n",
       "      <td>203091</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.073</td>\n",
       "      <td>0.098</td>\n",
       "      <td>55.148</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  interactome nodes self-loops   links connected components isolated nodes  \\\n",
       "0         SGI    83         60     256                   14             12   \n",
       "1           I  2677         38    3943                    4              0   \n",
       "2           U  6621       1638  203091                    1              0   \n",
       "3       I-LCC  2664          0    3895                    1              0   \n",
       "4       U-LCC  6621       1638  203091                    1              0   \n",
       "\n",
       "  average path length average clustering coefficient average degree diameter  \\\n",
       "0               3.872                          0.125          3.566        9   \n",
       "1                3.58                          0.014          2.917        9   \n",
       "2               3.073                          0.098         55.148        7   \n",
       "3                3.58                          0.014          2.923        9   \n",
       "4               3.073                          0.098         55.148        7   \n",
       "\n",
       "  radius centralization  \n",
       "0      1        0.00346  \n",
       "1      1          7e-05  \n",
       "2      1          5e-05  \n",
       "3      1          7e-05  \n",
       "4      1          5e-05  "
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measure = measure.rename({'nodeCount':'nodes', \n",
    "                          'nsl': 'self-loops', \n",
    "                          'ncc': 'connected components', \n",
    "                          'usn': 'isolated nodes', \n",
    "                          'avSpl': 'average path length', \n",
    "                          'cc': 'average clustering coefficient', \n",
    "                          'avNeighbors': 'average degree',\n",
    "                          'centralization': 'centralization'}, axis = 1)\n",
    "measure.to_csv('2.1_SGI_I_U_whole_network.csv',encoding='utf8',index=False)\n",
    "measure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local \n",
    "\n",
    "ii)\n",
    "- Node degree\n",
    "- Betweenness centrality\n",
    "- Eigenvector centrality\n",
    "- Closeness centrality\n",
    "- ratio Betweenness/Node degree\n",
    "\n",
    "\n",
    "#### I-LCC Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>Indegree</th>\n",
       "      <th>Outdegree</th>\n",
       "      <th>BetweennessCentrality</th>\n",
       "      <th>EigenvectorCentrality</th>\n",
       "      <th>ClosenessCentrality</th>\n",
       "      <th>ratio Betweenness/Nodedegree</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STAT3</td>\n",
       "      <td>74</td>\n",
       "      <td>84</td>\n",
       "      <td>0.006425</td>\n",
       "      <td>0.060636</td>\n",
       "      <td>0.340852</td>\n",
       "      <td>0.000041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>27</td>\n",
       "      <td>383</td>\n",
       "      <td>0.004870</td>\n",
       "      <td>0.313752</td>\n",
       "      <td>0.440842</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYC</td>\n",
       "      <td>16</td>\n",
       "      <td>508</td>\n",
       "      <td>0.003766</td>\n",
       "      <td>0.599906</td>\n",
       "      <td>0.446479</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HSPB1</td>\n",
       "      <td>17</td>\n",
       "      <td>250</td>\n",
       "      <td>0.002885</td>\n",
       "      <td>0.141267</td>\n",
       "      <td>0.367162</td>\n",
       "      <td>0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>15</td>\n",
       "      <td>44</td>\n",
       "      <td>0.002366</td>\n",
       "      <td>0.051102</td>\n",
       "      <td>0.354046</td>\n",
       "      <td>0.000040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDK6</td>\n",
       "      <td>22</td>\n",
       "      <td>62</td>\n",
       "      <td>0.002269</td>\n",
       "      <td>0.050420</td>\n",
       "      <td>0.405242</td>\n",
       "      <td>0.000027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PTPN11</td>\n",
       "      <td>71</td>\n",
       "      <td>40</td>\n",
       "      <td>0.002179</td>\n",
       "      <td>0.021083</td>\n",
       "      <td>0.536232</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MAPK8</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.001950</td>\n",
       "      <td>0.030569</td>\n",
       "      <td>0.280262</td>\n",
       "      <td>0.000390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RUNX1</td>\n",
       "      <td>25</td>\n",
       "      <td>24</td>\n",
       "      <td>0.001766</td>\n",
       "      <td>0.062564</td>\n",
       "      <td>0.437888</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FHL2</td>\n",
       "      <td>47</td>\n",
       "      <td>29</td>\n",
       "      <td>0.001695</td>\n",
       "      <td>0.044542</td>\n",
       "      <td>0.525773</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PPP2CA</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001268</td>\n",
       "      <td>0.043269</td>\n",
       "      <td>0.289398</td>\n",
       "      <td>0.000211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BCL2</td>\n",
       "      <td>15</td>\n",
       "      <td>43</td>\n",
       "      <td>0.001263</td>\n",
       "      <td>0.024805</td>\n",
       "      <td>0.299621</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EHMT2</td>\n",
       "      <td>51</td>\n",
       "      <td>18</td>\n",
       "      <td>0.001164</td>\n",
       "      <td>0.022917</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MET</td>\n",
       "      <td>9</td>\n",
       "      <td>23</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.011880</td>\n",
       "      <td>0.306561</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KMT2A</td>\n",
       "      <td>24</td>\n",
       "      <td>31</td>\n",
       "      <td>0.001037</td>\n",
       "      <td>0.058390</td>\n",
       "      <td>0.548077</td>\n",
       "      <td>0.000019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCEA2</td>\n",
       "      <td>16</td>\n",
       "      <td>27</td>\n",
       "      <td>0.001030</td>\n",
       "      <td>0.005825</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HSP90AA1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.001025</td>\n",
       "      <td>0.050688</td>\n",
       "      <td>0.320379</td>\n",
       "      <td>0.000103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EP300</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.047123</td>\n",
       "      <td>0.378641</td>\n",
       "      <td>0.000103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CD44</td>\n",
       "      <td>10</td>\n",
       "      <td>54</td>\n",
       "      <td>0.000837</td>\n",
       "      <td>0.019558</td>\n",
       "      <td>0.298177</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RUNX1T1</td>\n",
       "      <td>14</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000814</td>\n",
       "      <td>0.015013</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name  Indegree  Outdegree  BetweennessCentrality  \\\n",
       "0      STAT3        74         84               0.006425   \n",
       "1       NPM1        27        383               0.004870   \n",
       "2        MYC        16        508               0.003766   \n",
       "3      HSPB1        17        250               0.002885   \n",
       "4      CEBPA        15         44               0.002366   \n",
       "5       CDK6        22         62               0.002269   \n",
       "6     PTPN11        71         40               0.002179   \n",
       "7      MAPK8         3          2               0.001950   \n",
       "8      RUNX1        25         24               0.001766   \n",
       "9       FHL2        47         29               0.001695   \n",
       "10    PPP2CA         5          1               0.001268   \n",
       "11      BCL2        15         43               0.001263   \n",
       "12     EHMT2        51         18               0.001164   \n",
       "13       MET         9         23               0.001155   \n",
       "14     KMT2A        24         31               0.001037   \n",
       "15     TCEA2        16         27               0.001030   \n",
       "16  HSP90AA1         5          5               0.001025   \n",
       "17     EP300         6          3               0.000931   \n",
       "18      CD44        10         54               0.000837   \n",
       "19   RUNX1T1        14         27               0.000814   \n",
       "\n",
       "    EigenvectorCentrality  ClosenessCentrality  ratio Betweenness/Nodedegree  \n",
       "0                0.060636             0.340852                      0.000041  \n",
       "1                0.313752             0.440842                      0.000012  \n",
       "2                0.599906             0.446479                      0.000007  \n",
       "3                0.141267             0.367162                      0.000011  \n",
       "4                0.051102             0.354046                      0.000040  \n",
       "5                0.050420             0.405242                      0.000027  \n",
       "6                0.021083             0.536232                      0.000020  \n",
       "7                0.030569             0.280262                      0.000390  \n",
       "8                0.062564             0.437888                      0.000036  \n",
       "9                0.044542             0.525773                      0.000022  \n",
       "10               0.043269             0.289398                      0.000211  \n",
       "11               0.024805             0.299621                      0.000022  \n",
       "12               0.022917             0.875000                      0.000017  \n",
       "13               0.011880             0.306561                      0.000036  \n",
       "14               0.058390             0.548077                      0.000019  \n",
       "15               0.005825             1.000000                      0.000024  \n",
       "16               0.050688             0.320379                      0.000103  \n",
       "17               0.047123             0.378641                      0.000103  \n",
       "18               0.019558             0.298177                      0.000013  \n",
       "19               0.015013             0.571429                      0.000020  "
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.1_Data_analysis'\n",
    "I_LCC = pd.read_csv(os.path.join(path,'2.1. I_LCC default  node.csv'))\n",
    "I_LCC = I_LCC[['name', 'Indegree', 'Outdegree', 'BetweennessCentrality', 'ClosenessCentrality']]\n",
    "I_LCC['ratio Betweenness/Nodedegree'] = I_LCC['BetweennessCentrality'] / ( I_LCC['Indegree'] + I_LCC['Outdegree'])\n",
    "# Eigenvector Centrality --------------------\n",
    "eigen = nx.eigenvector_centrality_numpy(G_I_LCC)\n",
    "genes = list(eigen.keys())\n",
    "eigenvalue = list(eigen.values())\n",
    "eigen_centrality = pd.DataFrame(columns=['name', 'EigenvectorCentrality'])\n",
    "eigen_centrality['name'] = genes\n",
    "eigen_centrality['EigenvectorCentrality'] = eigenvalue\n",
    "\n",
    "I_LCC = pd.merge(I_LCC, eigen_centrality, on=['name'])\n",
    "I_LCC = I_LCC.sort_values(by=['BetweennessCentrality'], ascending=False)[:20]\n",
    "I_LCC = I_LCC[['name', 'Indegree', 'Outdegree', 'BetweennessCentrality', 'EigenvectorCentrality','ClosenessCentrality','ratio Betweenness/Nodedegree']]\n",
    "I_LCC = I_LCC.reset_index(drop=True)\n",
    "I_LCC.to_csv(os.path.join(path,'2.2_I_LCC_Highest_20_betweeness.csv'),encoding='utf8',index=False)\n",
    "I_LCC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### U-LCC Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>Indegree</th>\n",
       "      <th>Outdegree</th>\n",
       "      <th>BetweennessCentrality</th>\n",
       "      <th>EigenvectorCentrality</th>\n",
       "      <th>ClosenessCentrality</th>\n",
       "      <th>ratio Betweenness/Nodedegree</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYC</td>\n",
       "      <td>266</td>\n",
       "      <td>2190</td>\n",
       "      <td>0.073403</td>\n",
       "      <td>0.139624</td>\n",
       "      <td>0.568414</td>\n",
       "      <td>0.000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>APP</td>\n",
       "      <td>830</td>\n",
       "      <td>200</td>\n",
       "      <td>0.045589</td>\n",
       "      <td>0.050414</td>\n",
       "      <td>0.433620</td>\n",
       "      <td>0.000044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>60</td>\n",
       "      <td>1547</td>\n",
       "      <td>0.026270</td>\n",
       "      <td>0.067963</td>\n",
       "      <td>0.524812</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>453</td>\n",
       "      <td>635</td>\n",
       "      <td>0.024469</td>\n",
       "      <td>0.074395</td>\n",
       "      <td>0.489375</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>280</td>\n",
       "      <td>629</td>\n",
       "      <td>0.019164</td>\n",
       "      <td>0.076411</td>\n",
       "      <td>0.465607</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HSPB1</td>\n",
       "      <td>209</td>\n",
       "      <td>436</td>\n",
       "      <td>0.016570</td>\n",
       "      <td>0.038566</td>\n",
       "      <td>0.454877</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ELAVL1</td>\n",
       "      <td>123</td>\n",
       "      <td>828</td>\n",
       "      <td>0.016001</td>\n",
       "      <td>0.060041</td>\n",
       "      <td>0.502967</td>\n",
       "      <td>0.000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>STAT3</td>\n",
       "      <td>250</td>\n",
       "      <td>255</td>\n",
       "      <td>0.015236</td>\n",
       "      <td>0.021401</td>\n",
       "      <td>0.440136</td>\n",
       "      <td>0.000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TRIM25</td>\n",
       "      <td>92</td>\n",
       "      <td>1199</td>\n",
       "      <td>0.014806</td>\n",
       "      <td>0.090511</td>\n",
       "      <td>0.518984</td>\n",
       "      <td>0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PTPN11</td>\n",
       "      <td>221</td>\n",
       "      <td>173</td>\n",
       "      <td>0.014308</td>\n",
       "      <td>0.015533</td>\n",
       "      <td>0.419621</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HSP90AA1</td>\n",
       "      <td>326</td>\n",
       "      <td>395</td>\n",
       "      <td>0.014196</td>\n",
       "      <td>0.059755</td>\n",
       "      <td>0.466975</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>UBC</td>\n",
       "      <td>305</td>\n",
       "      <td>510</td>\n",
       "      <td>0.012836</td>\n",
       "      <td>0.063396</td>\n",
       "      <td>0.461952</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>190</td>\n",
       "      <td>777</td>\n",
       "      <td>0.012465</td>\n",
       "      <td>0.083558</td>\n",
       "      <td>0.497690</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>151</td>\n",
       "      <td>732</td>\n",
       "      <td>0.012087</td>\n",
       "      <td>0.058782</td>\n",
       "      <td>0.495363</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>VCP</td>\n",
       "      <td>238</td>\n",
       "      <td>459</td>\n",
       "      <td>0.011302</td>\n",
       "      <td>0.062682</td>\n",
       "      <td>0.474505</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>244</td>\n",
       "      <td>414</td>\n",
       "      <td>0.010829</td>\n",
       "      <td>0.063135</td>\n",
       "      <td>0.485673</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>165</td>\n",
       "      <td>672</td>\n",
       "      <td>0.010459</td>\n",
       "      <td>0.060406</td>\n",
       "      <td>0.469840</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HNRNPL</td>\n",
       "      <td>109</td>\n",
       "      <td>593</td>\n",
       "      <td>0.009429</td>\n",
       "      <td>0.047365</td>\n",
       "      <td>0.457562</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>LPAR1</td>\n",
       "      <td>97</td>\n",
       "      <td>144</td>\n",
       "      <td>0.009428</td>\n",
       "      <td>0.001718</td>\n",
       "      <td>0.335533</td>\n",
       "      <td>0.000039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>20</td>\n",
       "      <td>137</td>\n",
       "      <td>0.009198</td>\n",
       "      <td>0.003553</td>\n",
       "      <td>0.373333</td>\n",
       "      <td>0.000059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name  Indegree  Outdegree  BetweennessCentrality  \\\n",
       "0        MYC       266       2190               0.073403   \n",
       "1        APP       830        200               0.045589   \n",
       "2       KRAS        60       1547               0.026270   \n",
       "3       TP53       453        635               0.024469   \n",
       "4       NPM1       280        629               0.019164   \n",
       "5      HSPB1       209        436               0.016570   \n",
       "6     ELAVL1       123        828               0.016001   \n",
       "7      STAT3       250        255               0.015236   \n",
       "8     TRIM25        92       1199               0.014806   \n",
       "9     PTPN11       221        173               0.014308   \n",
       "10  HSP90AA1       326        395               0.014196   \n",
       "11       UBC       305        510               0.012836   \n",
       "12     BRCA1       190        777               0.012465   \n",
       "13      EGFR       151        732               0.012087   \n",
       "14       VCP       238        459               0.011302   \n",
       "15    CTNNB1       244        414               0.010829   \n",
       "16      XPO1       165        672               0.010459   \n",
       "17    HNRNPL       109        593               0.009429   \n",
       "18     LPAR1        97        144               0.009428   \n",
       "19      FLT3        20        137               0.009198   \n",
       "\n",
       "    EigenvectorCentrality  ClosenessCentrality  ratio Betweenness/Nodedegree  \n",
       "0                0.139624             0.568414                      0.000030  \n",
       "1                0.050414             0.433620                      0.000044  \n",
       "2                0.067963             0.524812                      0.000016  \n",
       "3                0.074395             0.489375                      0.000022  \n",
       "4                0.076411             0.465607                      0.000021  \n",
       "5                0.038566             0.454877                      0.000026  \n",
       "6                0.060041             0.502967                      0.000017  \n",
       "7                0.021401             0.440136                      0.000030  \n",
       "8                0.090511             0.518984                      0.000011  \n",
       "9                0.015533             0.419621                      0.000036  \n",
       "10               0.059755             0.466975                      0.000020  \n",
       "11               0.063396             0.461952                      0.000016  \n",
       "12               0.083558             0.497690                      0.000013  \n",
       "13               0.058782             0.495363                      0.000014  \n",
       "14               0.062682             0.474505                      0.000016  \n",
       "15               0.063135             0.485673                      0.000016  \n",
       "16               0.060406             0.469840                      0.000012  \n",
       "17               0.047365             0.457562                      0.000013  \n",
       "18               0.001718             0.335533                      0.000039  \n",
       "19               0.003553             0.373333                      0.000059  "
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.1_Data_analysis'\n",
    "U_LCC = pd.read_csv(os.path.join(path, '2.1. U default  node.csv'))\n",
    "U_LCC = U_LCC.sort_values(by=['BetweennessCentrality'], ascending=False)[:20]\n",
    "U_LCC = U_LCC[['name', 'Indegree', 'Outdegree', 'BetweennessCentrality', 'ClosenessCentrality']]\n",
    "U_LCC['ratio Betweenness/Nodedegree'] = U_LCC['BetweennessCentrality'] / ( U_LCC['Indegree'] + U_LCC['Outdegree'])\n",
    "\n",
    "# Eigenvector Centrality --------------------\n",
    "eigen = nx.eigenvector_centrality_numpy(G_U)\n",
    "genes = list(eigen.keys())\n",
    "eigenvalue = list(eigen.values())\n",
    "eigen_centrality = pd.DataFrame(columns=['name', 'EigenvectorCentrality'])\n",
    "eigen_centrality['name'] = genes\n",
    "eigen_centrality['EigenvectorCentrality'] = eigenvalue\n",
    "\n",
    "U_LCC = pd.merge(U_LCC, eigen_centrality, on=['name'])\n",
    "U_LCC = U_LCC.sort_values(by=['BetweennessCentrality'], ascending=False)[:20]\n",
    "U_LCC = U_LCC[['name', 'Indegree', 'Outdegree', 'BetweennessCentrality', 'EigenvectorCentrality','ClosenessCentrality','ratio Betweenness/Nodedegree']]\n",
    "U_LCC = U_LCC.reset_index(drop=True)\n",
    "U_LCC.to_csv(os.path.join(path,'2.2_U_LCC_Highest_20_betweeness.csv'),encoding='utf8',index=False)\n",
    "U_LCC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2) Apply clustering methods for disease modules discovery\n",
    "\n",
    "Cluster I-LCC and U-LCC using the **MCL** algorithm to get the modules.\n",
    "\n",
    "Once you have clustered the networks, find modules with no. of nodes >= 10 in which seed genes are statistically overrepresented (p < 0.05) by applying a hypergeometric test: such modules will be the “**putative disease modules**”.\n",
    "\n",
    "Store the results for both U-LCC and I-LCC in tables including in each row: clustering algorithm used, module ID, no. of seed genes in the module, total no. of genes in each module, seed gene IDs, all gene IDs in the module, p-value.\n",
    "\n",
    "### I-LCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of genes found in this network: 2664\n",
      "The number of seed genes found in this network: 93\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUID</th>\n",
       "      <th>__mclCluster</th>\n",
       "      <th>name</th>\n",
       "      <th>selected</th>\n",
       "      <th>shared name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4076</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TIMP3</td>\n",
       "      <td>False</td>\n",
       "      <td>TIMP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4074</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GTPBP2</td>\n",
       "      <td>False</td>\n",
       "      <td>GTPBP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4072</td>\n",
       "      <td>1.0</td>\n",
       "      <td>KRTAP4-12</td>\n",
       "      <td>False</td>\n",
       "      <td>KRTAP4-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4069</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ELP3</td>\n",
       "      <td>False</td>\n",
       "      <td>ELP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4067</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LLGL2</td>\n",
       "      <td>False</td>\n",
       "      <td>LLGL2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SUID  __mclCluster       name  selected shared name\n",
       "0  4076           1.0      TIMP3     False       TIMP3\n",
       "1  4074           1.0     GTPBP2     False      GTPBP2\n",
       "2  4072           1.0  KRTAP4-12     False   KRTAP4-12\n",
       "3  4069           1.0       ELP3     False        ELP3\n",
       "4  4067           1.0      LLGL2     False       LLGL2"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "I_LCC_MCL = pd.read_csv(os.path.join(path,'2.2. I_LCC--clustered default  node.csv'))\n",
    "print('The number of genes found in this network:', len(I_LCC_MCL))\n",
    "print('The number of seed genes found in this network:', len(set(I_LCC_MCL['name'].values).intersection(seed_genes)))\n",
    "I_LCC_MCL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I-LCC\n",
      "The number of clusters having more than 10 genes: 15\n",
      "The number of genes in all clusters: 2664\n"
     ]
    }
   ],
   "source": [
    "print('I-LCC')\n",
    "clusters_I = {}\n",
    "for moduleID in range(1, 30):\n",
    "    current_df = I_LCC_MCL[I_LCC_MCL['__mclCluster'] == moduleID]\n",
    "    \n",
    "    if len(current_df) < 10: # only the cluster having at least 10 genes\n",
    "        break\n",
    "        \n",
    "    clusters_I[moduleID] = list(current_df.name.values)\n",
    "    \n",
    "num_clusters = len(clusters_I.keys())\n",
    "num_genes_I_LCC = len(I_LCC_MCL)\n",
    "print('The number of clusters having more than 10 genes: {}'.format(num_clusters))\n",
    "print('The number of genes in all clusters: {}'.format(num_genes_I_LCC))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### mapping gene-ID (used in 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18198\n"
     ]
    }
   ],
   "source": [
    "df_mapping =  pd.read_csv(os.path.join(path,'2.2. mapping.csv'))\n",
    "df_mapping.dropna(inplace=True)\n",
    "df_mapping['GeneID']=df_mapping['GeneID'].astype('int')\n",
    "mapping = dict(zip(df_mapping.Gene, df_mapping.GeneID))\n",
    "print(len(mapping)) #18194"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene      0\n",
      "GeneID    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# deleting the Nan\n",
    "print(df_mapping.isna().sum())\n",
    "del mapping['Uncharacterized'], mapping['SMAP'], mapping['Tcr-alpha'], mapping['cDNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = dict(zip(df_mapping.Gene, df_mapping.GeneID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of putative disease modules 6\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import hypergeom\n",
    "\n",
    "M = num_genes_I_LCC\n",
    "N = len(set(I_LCC_MCL['name'].values).intersection(seed_genes))\n",
    "output = {}\n",
    "\n",
    "for moduleID, genes in clusters_I.items():\n",
    "    module_genes = genes\n",
    "    module_genes_ID = [mapping[i] for i in module_genes]\n",
    "    n = len(module_genes_ID)\n",
    "\n",
    "    seed_genes_in_cluster = list(set(module_genes).intersection(seed_genes))\n",
    "    seed_genes_in_cluster_ID = [mapping[i] for i in seed_genes_in_cluster]\n",
    "    k = len(seed_genes_in_cluster_ID)\n",
    "    \n",
    "    p_value = hypergeom.sf(k, M, n, N)\n",
    "\n",
    "    output[moduleID]  = [seed_genes_in_cluster_ID, module_genes_ID, p_value]\n",
    "    \n",
    "putative_I_LCC = {}\n",
    "for i in list(output.keys()):\n",
    "        if output[i][2] < 0.05:\n",
    "            putative_I_LCC[i] = output[i]\n",
    "            \n",
    "print('the number of putative disease modules', len(putative_I_LCC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of genes found in this network: 6621\n",
      "The number of seed genes found in this network: 106\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUID</th>\n",
       "      <th>__mclCluster</th>\n",
       "      <th>name</th>\n",
       "      <th>selected</th>\n",
       "      <th>shared name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>223772</td>\n",
       "      <td>1.0</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>False</td>\n",
       "      <td>IL17A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>223744</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ADGRL1</td>\n",
       "      <td>False</td>\n",
       "      <td>ADGRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>223711</td>\n",
       "      <td>3.0</td>\n",
       "      <td>UPK3BL1</td>\n",
       "      <td>False</td>\n",
       "      <td>UPK3BL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>223701</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ABHD18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABHD18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>223689</td>\n",
       "      <td>34.0</td>\n",
       "      <td>SEPTIN8</td>\n",
       "      <td>False</td>\n",
       "      <td>SEPTIN8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUID  __mclCluster     name  selected shared name\n",
       "0  223772           1.0    IL17A     False       IL17A\n",
       "1  223744           3.0   ADGRL1     False      ADGRL1\n",
       "2  223711           3.0  UPK3BL1     False     UPK3BL1\n",
       "3  223701           3.0   ABHD18     False      ABHD18\n",
       "4  223689          34.0  SEPTIN8     False     SEPTIN8"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### U-LCC\n",
    "\n",
    "U_LCC_MCL = pd.read_csv(os.path.join(path,'2.2. U--clustered default  node.csv'))\n",
    "print('The number of genes found in this network:', len(U_LCC_MCL))\n",
    "print('The number of seed genes found in this network:', len(set(U_LCC_MCL['name'].values).intersection(seed_genes)))\n",
    "U_LCC_MCL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "U-LCC\n",
      "The number of clusters having more than 10 genes: 29\n",
      "The number of genes in all clusters: 6621\n"
     ]
    }
   ],
   "source": [
    "print('U-LCC')\n",
    "clusters_U = {}\n",
    "for moduleID in range(1, 30):\n",
    "    current_df = U_LCC_MCL[U_LCC_MCL['__mclCluster'] == moduleID]\n",
    "    \n",
    "    if len(current_df) < 10:\n",
    "        break\n",
    "        \n",
    "    clusters_U[moduleID] = list(current_df.name.values)\n",
    "        \n",
    "num_clusters = len(clusters_U.keys())\n",
    "num_genes_U_LCC = len(U_LCC_MCL)\n",
    "print('The number of clusters having more than 10 genes: {}'.format(num_clusters))\n",
    "print('The number of genes in all clusters: {}'.format(num_genes_U_LCC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of putative disease modules: 2\n"
     ]
    }
   ],
   "source": [
    "M = num_genes_U_LCC\n",
    "N = len(set(U_LCC_MCL['name'].values).intersection(seed_genes))\n",
    "output = {}\n",
    "\n",
    "for moduleID, genes in clusters_U.items():\n",
    "    module_genes = genes\n",
    "    module_genes_ID = [mapping[i] for i in module_genes]\n",
    "    n = len(module_genes_ID)\n",
    "\n",
    "    seed_genes_in_cluster = list(set(module_genes).intersection(seed_genes))\n",
    "    seed_genes_in_cluster_ID = [mapping[i] for i in seed_genes_in_cluster]\n",
    "    k = len(seed_genes_in_cluster_ID)\n",
    "    \n",
    "    p_value = hypergeom.sf(k, M, n, N)\n",
    "\n",
    "    output[moduleID]  = [seed_genes_in_cluster_ID, module_genes_ID, p_value]\n",
    "    \n",
    "putative_U_LCC = {}\n",
    "for i in list(output.keys()):\n",
    "        if output[i][2] < 0.05:\n",
    "            putative_U_LCC[i] = output[i]\n",
    "print('the number of putative disease modules:', len(putative_U_LCC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactome</th>\n",
       "      <th>algorithm</th>\n",
       "      <th>module ID</th>\n",
       "      <th>no. of seed genes</th>\n",
       "      <th>total no. of genes</th>\n",
       "      <th>ratio seed/total</th>\n",
       "      <th>seed gene IDs</th>\n",
       "      <th>all gene IDs</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>904</td>\n",
       "      <td>0.05973</td>\n",
       "      <td>6840,4869,2969,5997,7280,8623,864,4629,6279,10...</td>\n",
       "      <td>7078,54676,83755,55140,3993,64854,29934,23196,...</td>\n",
       "      <td>3.63107e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "      <td>0.08333</td>\n",
       "      <td>3418,307,308</td>\n",
       "      <td>4357,6609,310,307,131578,49860,9374,10319,2222...</td>\n",
       "      <td>0.0349096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>0.09091</td>\n",
       "      <td>1978,6446</td>\n",
       "      <td>3182,5578,9113,472,1977,6788,699,208,4043,2317...</td>\n",
       "      <td>0.03923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.1</td>\n",
       "      <td>824</td>\n",
       "      <td>637,471,2762,9601,7284,6635,4287,5798,9474,824</td>\n",
       "      <td>0.0452104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.1</td>\n",
       "      <td>53826</td>\n",
       "      <td>5501,221074,56928,9564,722,55142,79939,53826,4...</td>\n",
       "      <td>0.0452104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.1</td>\n",
       "      <td>4893</td>\n",
       "      <td>5910,6224,3337,5294,673,65244,5290,79621,4893,...</td>\n",
       "      <td>0.0452104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>U-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>35</td>\n",
       "      <td>0.05714</td>\n",
       "      <td>3417,2026</td>\n",
       "      <td>51166,84173,441282,140458,654364,231,9380,5019...</td>\n",
       "      <td>0.0180074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>U-LCC</td>\n",
       "      <td>Marcov</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>0.07407</td>\n",
       "      <td>4330,1052</td>\n",
       "      <td>4330,25932,23234,55689,7264,252969,51056,55818...</td>\n",
       "      <td>0.00882408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  interactome algorithm module ID no. of seed genes total no. of genes  \\\n",
       "0       I-LCC    Marcov         1                54                904   \n",
       "1       I-LCC    Marcov         4                 3                 36   \n",
       "2       I-LCC    Marcov         6                 2                 22   \n",
       "3       I-LCC    Marcov        13                 1                 10   \n",
       "4       I-LCC    Marcov        14                 1                 10   \n",
       "5       I-LCC    Marcov        15                 1                 10   \n",
       "6       U-LCC    Marcov        15                 2                 35   \n",
       "7       U-LCC    Marcov        24                 2                 27   \n",
       "\n",
       "  ratio seed/total                                      seed gene IDs  \\\n",
       "0          0.05973  6840,4869,2969,5997,7280,8623,864,4629,6279,10...   \n",
       "1          0.08333                                       3418,307,308   \n",
       "2          0.09091                                          1978,6446   \n",
       "3              0.1                                                824   \n",
       "4              0.1                                              53826   \n",
       "5              0.1                                               4893   \n",
       "6          0.05714                                          3417,2026   \n",
       "7          0.07407                                          4330,1052   \n",
       "\n",
       "                                        all gene IDs      p-value  \n",
       "0  7078,54676,83755,55140,3993,64854,29934,23196,...  3.63107e-07  \n",
       "1  4357,6609,310,307,131578,49860,9374,10319,2222...    0.0349096  \n",
       "2  3182,5578,9113,472,1977,6788,699,208,4043,2317...      0.03923  \n",
       "3     637,471,2762,9601,7284,6635,4287,5798,9474,824    0.0452104  \n",
       "4  5501,221074,56928,9564,722,55142,79939,53826,4...    0.0452104  \n",
       "5  5910,6224,3337,5294,673,65244,5290,79621,4893,...    0.0452104  \n",
       "6  51166,84173,441282,140458,654364,231,9380,5019...    0.0180074  \n",
       "7  4330,25932,23234,55689,7264,252969,51056,55818...   0.00882408  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_putative = pd.DataFrame(columns=['interactome',  'algorithm', 'module ID', \n",
    "                                    'no. of seed genes'  ,'total no. of genes','ratio seed/total',\n",
    "                                    'seed gene IDs', 'all gene IDs', 'p-value'])\n",
    "\n",
    "\n",
    "count = 0\n",
    "# I-LCC ------------------------------------------------------------------------------------\n",
    "for moduleID,v in putative_I_LCC.items():\n",
    "    df_putative.set_value(count, 'interactome', 'I-LCC')\n",
    "    df_putative.set_value(count, 'algorithm', 'Marcov')\n",
    "    df_putative.set_value(count, 'module ID', moduleID)\n",
    "    df_putative.set_value(count, 'no. of seed genes', len(v[0]))\n",
    "    df_putative.set_value(count, 'total no. of genes', len(v[1]))\n",
    "    df_putative.set_value(count, 'ratio seed/total', round(len(v[0])/len(v[1]),5))\n",
    "    df_putative.set_value(count, 'seed gene IDs', ','.join(str(int(v)) for v in v[0]))\n",
    "    df_putative.set_value(count, 'all gene IDs', ','.join(str(int(v)) for v in v[1]))\n",
    "    df_putative.set_value(count, 'p-value', v[2])\n",
    "    count += 1\n",
    "    \n",
    "# U-LCC ------------------------------------------------------------------------------------\n",
    "for moduleID,v in putative_U_LCC.items():\n",
    "    df_putative.set_value(count, 'interactome', 'U-LCC')\n",
    "    df_putative.set_value(count, 'algorithm', 'Marcov')\n",
    "    df_putative.set_value(count, 'module ID', moduleID)\n",
    "    df_putative.set_value(count, 'no. of seed genes', len(v[0]))\n",
    "    df_putative.set_value(count, 'total no. of genes', len(v[1]))\n",
    "    df_putative.set_value(count, 'ratio seed/total', round(len(v[0])/len(v[1]),5))\n",
    "    df_putative.set_value(count, 'seed gene IDs', v[0])\n",
    "    df_putative.set_value(count, 'seed gene IDs', ','.join(str(int(v)) for v in v[0]))\n",
    "    df_putative.set_value(count, 'all gene IDs', ','.join(str(int(v)) for v in v[1]))\n",
    "    df_putative.set_value(count, 'p-value', v[2])\n",
    "    count += 1\n",
    "\n",
    "df_putative.to_csv(os.path.join(path,'2.2. putative.csv'),index=False)\n",
    "df_putative"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3) Carry on an enrichment analysis on the disease modules\n",
    "\n",
    "Find overrepresented GO categories (limit to first ten) and overrepresented pathways (limit to first ten) for the genes belonging to each putative disease module.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting Gene List for I_LCC and U_LCC\n",
    "\n",
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.3 EnrichmentAnalysis'\n",
    "genes_I_LCC_1 = df_putative.iloc[0]['all gene IDs'].split(',')\n",
    "genes_I_LCC_1 = [mapping2[int(i)] for i in genes_I_LCC_1]\n",
    "pd.Series(genes_I_LCC_1).to_csv(os.path.join(path,'2.3. genes_I_LCC.csv'), index = False)\n",
    "\n",
    "genes_U_LCC_1 = df_putative.iloc[6]['all gene IDs'].split(',')\n",
    "genes_U_LCC_1 = [mapping2[int(i)] for i in genes_U_LCC_1]\n",
    "pd.Series(genes_U_LCC_1).to_csv(os.path.join(path,'2.3. genes_U_LCC.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathways in cancer</td>\n",
       "      <td>94/530</td>\n",
       "      <td>7.483322e-31</td>\n",
       "      <td>2.304863e-28</td>\n",
       "      <td>3.923860</td>\n",
       "      <td>272.188234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>39/124</td>\n",
       "      <td>1.067451e-22</td>\n",
       "      <td>1.643874e-20</td>\n",
       "      <td>6.958321</td>\n",
       "      <td>352.032605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Viral carcinogenesis</td>\n",
       "      <td>46/201</td>\n",
       "      <td>3.400096e-20</td>\n",
       "      <td>3.490765e-18</td>\n",
       "      <td>5.063180</td>\n",
       "      <td>226.971703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Human T-cell leukemia virus 1 infection</td>\n",
       "      <td>48/219</td>\n",
       "      <td>3.556681e-20</td>\n",
       "      <td>2.738644e-18</td>\n",
       "      <td>4.849073</td>\n",
       "      <td>217.155408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hepatitis B</td>\n",
       "      <td>41/163</td>\n",
       "      <td>9.399751e-20</td>\n",
       "      <td>5.790247e-18</td>\n",
       "      <td>5.564906</td>\n",
       "      <td>243.804192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI3K-Akt signaling pathway</td>\n",
       "      <td>61/354</td>\n",
       "      <td>1.164505e-19</td>\n",
       "      <td>5.977794e-18</td>\n",
       "      <td>3.812309</td>\n",
       "      <td>166.204568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Human cytomegalovirus infection</td>\n",
       "      <td>48/225</td>\n",
       "      <td>1.180034e-19</td>\n",
       "      <td>5.192148e-18</td>\n",
       "      <td>4.719764</td>\n",
       "      <td>205.704184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FoxO signaling pathway</td>\n",
       "      <td>36/132</td>\n",
       "      <td>9.330817e-19</td>\n",
       "      <td>3.592364e-17</td>\n",
       "      <td>6.033789</td>\n",
       "      <td>250.497552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Human papillomavirus infection</td>\n",
       "      <td>57/330</td>\n",
       "      <td>1.701279e-18</td>\n",
       "      <td>5.822154e-17</td>\n",
       "      <td>3.821400</td>\n",
       "      <td>156.353153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>37/155</td>\n",
       "      <td>3.841833e-17</td>\n",
       "      <td>1.183285e-15</td>\n",
       "      <td>5.281188</td>\n",
       "      <td>199.618311</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Term Overlap       P-value  \\\n",
       "0                       Pathways in cancer  94/530  7.483322e-31   \n",
       "1                               Cell cycle  39/124  1.067451e-22   \n",
       "2                     Viral carcinogenesis  46/201  3.400096e-20   \n",
       "3  Human T-cell leukemia virus 1 infection  48/219  3.556681e-20   \n",
       "4                              Hepatitis B  41/163  9.399751e-20   \n",
       "5               PI3K-Akt signaling pathway  61/354  1.164505e-19   \n",
       "6          Human cytomegalovirus infection  48/225  1.180034e-19   \n",
       "7                   FoxO signaling pathway  36/132  9.330817e-19   \n",
       "8           Human papillomavirus infection  57/330  1.701279e-18   \n",
       "9                              Hepatitis C  37/155  3.841833e-17   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0      2.304863e-28    3.923860      272.188234  \n",
       "1      1.643874e-20    6.958321      352.032605  \n",
       "2      3.490765e-18    5.063180      226.971703  \n",
       "3      2.738644e-18    4.849073      217.155408  \n",
       "4      5.790247e-18    5.564906      243.804192  \n",
       "5      5.977794e-18    3.812309      166.204568  \n",
       "6      5.192148e-18    4.719764      205.704184  \n",
       "7      3.592364e-17    6.033789      250.497552  \n",
       "8      5.822154e-17    3.821400      156.353153  \n",
       "9      1.183285e-15    5.281188      199.618311  "
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.3 EnrichmentAnalysis\\Enrichr_I_LCC'\n",
    "\n",
    "# GO Analysis I_LCC\n",
    "\n",
    "# GO Biological Process 2018\n",
    "I_LCC_BP = pd.read_csv(os.path.join(path , 'GO_Biological_Process_2018_table.txt'),sep='\\t')\n",
    "I_LCC_BP = I_LCC_BP[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "I_LCC_BP.to_csv(os.path.join(path ,'2.3_a_GO_Biological_Process_2018_table.csv'),encoding='utf8',index=False)\n",
    "I_LCC_BP.head(10)\n",
    "\n",
    "#GO_Molecular_Function_2018\n",
    "I_LCC_MF = pd.read_csv(os.path.join(path , 'GO_Molecular_Function_2018_table.txt'),sep='\\t')\n",
    "I_LCC_MF = I_LCC_MF[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "I_LCC_MF.to_csv(os.path.join(path ,'2.3_a_GO_Molecular_Function_2018_table.csv'),encoding='utf8',index=False)\n",
    "I_LCC_MF.head(10)\n",
    "\n",
    "#GO_Cellular_Component_2018\n",
    "I_LCC_CC = pd.read_csv(os.path.join(path , 'GO_Cellular_Component_2018_table.txt'),sep='\\t')\n",
    "I_LCC_CC = I_LCC_CC[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "I_LCC_CC.to_csv(os.path.join(path ,'2.3_a_GO_Cellular_Component_2018_table.csv'),encoding='utf8',index=False)\n",
    "I_LCC_CC.head(10)\n",
    "\n",
    "# Pathway Analysis I_LCC\n",
    "\n",
    "# KEGG 2019 HUMAN\n",
    "kegg_I_LCC_enrichr = pd.read_csv(os.path.join(path,'KEGG_2019_Human_table.txt'),sep='\\t')\n",
    "kegg_I_LCC_enrichr = kegg_I_LCC_enrichr[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "kegg_I_LCC_enrichr.to_csv(os.path.join(path ,'2.3_a_KEGG_2019_Human_table.csv'),encoding='utf8',index=False)\n",
    "kegg_I_LCC_enrichr.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>4/68</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.001820</td>\n",
       "      <td>33.613445</td>\n",
       "      <td>404.671558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIF-1 signaling pathway</td>\n",
       "      <td>4/100</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.004211</td>\n",
       "      <td>22.857143</td>\n",
       "      <td>240.158046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glutathione metabolism</td>\n",
       "      <td>3/56</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.013111</td>\n",
       "      <td>30.612245</td>\n",
       "      <td>274.463473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Butanoate metabolism</td>\n",
       "      <td>2/28</td>\n",
       "      <td>0.001093</td>\n",
       "      <td>0.084153</td>\n",
       "      <td>40.816327</td>\n",
       "      <td>278.323580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "      <td>2/30</td>\n",
       "      <td>0.001255</td>\n",
       "      <td>0.077304</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>254.501736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>2/30</td>\n",
       "      <td>0.001255</td>\n",
       "      <td>0.064420</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>254.501736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>2/40</td>\n",
       "      <td>0.002226</td>\n",
       "      <td>0.097930</td>\n",
       "      <td>28.571429</td>\n",
       "      <td>174.505437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "      <td>2/42</td>\n",
       "      <td>0.002451</td>\n",
       "      <td>0.094380</td>\n",
       "      <td>27.210884</td>\n",
       "      <td>163.566858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fatty acid degradation</td>\n",
       "      <td>2/44</td>\n",
       "      <td>0.002688</td>\n",
       "      <td>0.091974</td>\n",
       "      <td>25.974026</td>\n",
       "      <td>153.743529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Valine, leucine and isoleucine degradation</td>\n",
       "      <td>2/48</td>\n",
       "      <td>0.003191</td>\n",
       "      <td>0.098270</td>\n",
       "      <td>23.809524</td>\n",
       "      <td>136.846417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Term Overlap   P-value  \\\n",
       "0                Glycolysis / Gluconeogenesis    4/68  0.000006   \n",
       "1                     HIF-1 signaling pathway   4/100  0.000027   \n",
       "2                      Glutathione metabolism    3/56  0.000128   \n",
       "3                        Butanoate metabolism    2/28  0.001093   \n",
       "4                   Citrate cycle (TCA cycle)    2/30  0.001255   \n",
       "5     Glyoxylate and dicarboxylate metabolism    2/30  0.001255   \n",
       "6    Glycine, serine and threonine metabolism    2/40  0.002226   \n",
       "7                       Tryptophan metabolism    2/42  0.002451   \n",
       "8                      Fatty acid degradation    2/44  0.002688   \n",
       "9  Valine, leucine and isoleucine degradation    2/48  0.003191   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \n",
       "0          0.001820   33.613445      404.671558  \n",
       "1          0.004211   22.857143      240.158046  \n",
       "2          0.013111   30.612245      274.463473  \n",
       "3          0.084153   40.816327      278.323580  \n",
       "4          0.077304   38.095238      254.501736  \n",
       "5          0.064420   38.095238      254.501736  \n",
       "6          0.097930   28.571429      174.505437  \n",
       "7          0.094380   27.210884      163.566858  \n",
       "8          0.091974   25.974026      153.743529  \n",
       "9          0.098270   23.809524      136.846417  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GO Analysis for U_LCC\n",
    "\n",
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\2.3 EnrichmentAnalysis\\Enrichr_U_LCC'\n",
    "\n",
    "U_LCC_BP = pd.read_csv(os.path.join(path , 'GO_Biological_Process_2018_table.txt'),sep='\\t')\n",
    "U_LCC_BP = U_LCC_BP[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "U_LCC_BP.to_csv(os.path.join(path ,'2.3_b_GO_Biological_Process_2018_table.csv'),encoding='utf8',index=False)\n",
    "U_LCC_BP.head(10)\n",
    "\n",
    "U_LCC_MF = pd.read_csv(os.path.join(path , 'GO_Molecular_Function_2018_table.txt'),sep='\\t')\n",
    "U_LCC_MF = U_LCC_MF[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "U_LCC_MF.to_csv(os.path.join(path ,'2.3_b_GO_Molecular_Function_2018_table.csv'),encoding='utf8',index=False)\n",
    "U_LCC_MF.head(10)\n",
    "\n",
    "U_LCC_CC = pd.read_csv(os.path.join(path , 'GO_Cellular_Component_2018_table.txt'),sep='\\t')\n",
    "U_LCC_CC = U_LCC_CC[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "U_LCC_CC.to_csv(os.path.join(path ,'2.3_b_GO_Cellular_Component_2018_table.csv'),encoding='utf8',index=False)\n",
    "U_LCC_CC.head(10)\n",
    "\n",
    "# Pathway Analysis\n",
    "\n",
    "# KEGG 2019 HUMAN\n",
    "kegg_U_LCC_enrichr = pd.read_csv(os.path.join(path,'KEGG_2019_Human_table.txt'),sep='\\t')\n",
    "kegg_U_LCC_enrichr = kegg_U_LCC_enrichr[['Term','Overlap','P-value','Adjusted P-value','Odds Ratio','Combined Score']]\n",
    "kegg_U_LCC_enrichr.to_csv(os.path.join(path ,'2.3_b_KEGG_2019_Human_table.csv'),encoding='utf8',index=False)\n",
    "kegg_U_LCC_enrichr.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 2.4) Find putative disease genes using the DIAMOnD tool\n",
    "\n",
    "Using the tool DIAMOnD, compute the putative disease protein list using as reference interactome (“network_file”) the whole BioGrid interactome already used to collect PPIs. As “seed_file” use your seed gene list, limit the number of putative disease proteins (“n”) to 200, and omit the “alpha” parameter (it will be set by default to 1).\n",
    "\n",
    "Software and instruction for [DIAMOnD](https://github.com/barabasilab/DIAMOnD)\n",
    "\n",
    "Find overrepresented GO categories (limit to first ten) and overrepresented pathways (limit to first ten) of such 200 newly found genes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Entrez Gene Interactor A Entrez Gene Interactor B\n",
       "0                     6416                     2318\n",
       "1                    84665                       88\n",
       "2                       90                     2339\n",
       "3                     2624                     5371\n",
       "4                     6118                     6774"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_biogrid = pd.read_csv('BIOGRID-ALL-3.5.180.tab2.txt', sep='\\t')\n",
    "df_biogrid = df_biogrid.loc[(df_biogrid['Organism Interactor A'] == 9606) & (df_biogrid['Organism Interactor B'] == 9606)]\n",
    "network_file = df_biogrid[['Entrez Gene Interactor A', 'Entrez Gene Interactor B']]\n",
    "\n",
    "network_file.to_csv('DIAMOnD_network_file.txt', index = False, header=False)\n",
    "network_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracting the seed list to perform diamond analysis\n",
    "pd.Series(seed_geneIDs).to_csv('DIAMOnD_seed_file.txt', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a map dictionary for reverse look-up\n",
    "mapping2 = dict([(value, key) for key, value in mapping.items()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rank</th>\n",
       "      <th>DIAMOnD_node</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>5925</td>\n",
       "      <td>RB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>5894</td>\n",
       "      <td>RAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>5371</td>\n",
       "      <td>PML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>84662</td>\n",
       "      <td>GLIS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1969</td>\n",
       "      <td>EPHA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>405</td>\n",
       "      <td>ARNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>1030</td>\n",
       "      <td>CDKN2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>5607</td>\n",
       "      <td>MAP2K5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>2264</td>\n",
       "      <td>FGFR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>1019</td>\n",
       "      <td>CDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>5606</td>\n",
       "      <td>MAP2K3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>2064</td>\n",
       "      <td>ERBB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>6794</td>\n",
       "      <td>STK11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>2911</td>\n",
       "      <td>GRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>1031</td>\n",
       "      <td>CDKN2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>11014</td>\n",
       "      <td>KDELR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>4771</td>\n",
       "      <td>NF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>26524</td>\n",
       "      <td>LATS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>8678</td>\n",
       "      <td>BECN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>207</td>\n",
       "      <td>AKT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>7015</td>\n",
       "      <td>TERT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>11186</td>\n",
       "      <td>RASSF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>51343</td>\n",
       "      <td>FZR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>8463</td>\n",
       "      <td>TEAD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>898</td>\n",
       "      <td>CCNE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>27</td>\n",
       "      <td>1029</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>28</td>\n",
       "      <td>369</td>\n",
       "      <td>ARAF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>29</td>\n",
       "      <td>4217</td>\n",
       "      <td>MAP3K5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>30</td>\n",
       "      <td>23624</td>\n",
       "      <td>CBLC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Rank  DIAMOnD_node    Gene\n",
       "0      1          5925     RB1\n",
       "1      2          5894    RAF1\n",
       "2      3          5371     PML\n",
       "3      4         84662   GLIS2\n",
       "4      5          1969   EPHA2\n",
       "5      6           405    ARNT\n",
       "6      7          1030  CDKN2B\n",
       "7      8          5607  MAP2K5\n",
       "8      9          2264   FGFR4\n",
       "9     10          1019    CDK4\n",
       "10    11          5156  PDGFRA\n",
       "11    12          5606  MAP2K3\n",
       "12    13          2064   ERBB2\n",
       "13    14          6794   STK11\n",
       "14    15          2911    GRM1\n",
       "15    16          1031  CDKN2C\n",
       "16    17         11014  KDELR2\n",
       "17    18          4771     NF2\n",
       "18    19         26524   LATS2\n",
       "19    20          8678   BECN1\n",
       "20    21           207    AKT1\n",
       "21    22          7015    TERT\n",
       "22    23         11186  RASSF1\n",
       "23    24         51343    FZR1\n",
       "24    25          8463   TEAD2\n",
       "25    26           898   CCNE1\n",
       "26    27          1029  CDKN2A\n",
       "27    28           369    ARAF\n",
       "28    29          4217  MAP3K5\n",
       "29    30         23624    CBLC"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The First 30 Genes from Diamond Analysis\n",
    "path = r'C:\\Users\\alevi\\DS_HOME\\BioInformatics\\Project2\\DIAMOnD'\n",
    "df_diamond = pd.read_csv(os.path.join(path,'first_200_added_nodes_weight_1.txt'), sep='\\t')\n",
    "df_diamond.rename(columns={'#rank':'Rank'},inplace=True)\n",
    "df_diamond['Gene'] = df_diamond['DIAMOnD_node'].map(mapping2)\n",
    "df_diamond['Gene'].to_csv(os.path.join(path,'2.4_diamond_genes.csv'),encoding='utf8',index=False)\n",
    "df_diamond.to_csv(os.path.join(path,'2.4_First_30_genes.csv'),encoding='utf8',index=False)\n",
    "df_diamond.head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'> ---- THE  END ---- </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
